From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sun, 27 Jul 2025 14:47:04 +0200
Subject: [PATCH 092/225] WIP: media: i2c: add OV64B40 sensor driver
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The OV64B40 sensor is similar to the OV64A40 sensor, but not entirely
compatible with it. Add a new driver for the sensor with updated
initialization sequences. Some code (e.g. the subsampling configuration)
was removed due to the lack of documentation.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/media/i2c/Kconfig   |   19 +
 drivers/media/i2c/Makefile  |    2 +
 drivers/media/i2c/ov64b40.c | 2901 +++++++++++++++++++++++++++++++++++
 3 files changed, 2922 insertions(+)
 create mode 100644 drivers/media/i2c/ov64b40.c

diff --git a/drivers/media/i2c/Kconfig b/drivers/media/i2c/Kconfig
index 8f2ba4121..dd86ac983 100644
--- a/drivers/media/i2c/Kconfig
+++ b/drivers/media/i2c/Kconfig
@@ -630,6 +630,25 @@ config VIDEO_OV64A40
 	  To compile this driver as a module, choose M here: the
 	  module will be called ov64a40.
 
+config VIDEO_OV64B40
+	tristate "OmniVision OV64B40 sensor support"
+	select V4L2_CCI_I2C
+	help
+	  This is a Video4Linux2 sensor driver for the OmniVision
+	  OV64B40 camera.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called ov64b40.
+
+config VIDEO_OV6650
+	tristate "OmniVision OV6650 sensor support"
+	help
+	  This is a Video4Linux2 sensor driver for the OmniVision
+	  OV6650 camera.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called ov6650.
+
 config VIDEO_OV7251
 	tristate "OmniVision OV7251 sensor support"
 	help
diff --git a/drivers/media/i2c/Makefile b/drivers/media/i2c/Makefile
index 90b276a74..178822235 100644
--- a/drivers/media/i2c/Makefile
+++ b/drivers/media/i2c/Makefile
@@ -111,6 +111,8 @@ obj-$(CONFIG_VIDEO_OV5693) += ov5693.o
 obj-$(CONFIG_VIDEO_OV5695) += ov5695.o
 obj-$(CONFIG_VIDEO_OV6211) += ov6211.o
 obj-$(CONFIG_VIDEO_OV64A40) += ov64a40.o
+obj-$(CONFIG_VIDEO_OV64B40) += ov64b40.o
+obj-$(CONFIG_VIDEO_OV6650) += ov6650.o
 obj-$(CONFIG_VIDEO_OV7251) += ov7251.o
 obj-$(CONFIG_VIDEO_OV7640) += ov7640.o
 obj-$(CONFIG_VIDEO_OV7670) += ov7670.o
diff --git a/drivers/media/i2c/ov64b40.c b/drivers/media/i2c/ov64b40.c
new file mode 100644
index 000000000..620572f44
--- /dev/null
+++ b/drivers/media/i2c/ov64b40.c
@@ -0,0 +1,2901 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * V4L2 sensor driver for OmniVision OV64B40
+ *
+ * Except for the register initialization, this driver is a copy of ov64a40.c,
+ * Copyright (C) 2023 Ideas On Board Oy
+ * Copyright (C) 2023 Arducam
+ */
+
+#include <linux/clk.h>
+#include <linux/delay.h>
+#include <linux/gpio/consumer.h>
+#include <linux/i2c.h>
+#include <linux/mod_devicetable.h>
+#include <linux/module.h>
+#include <linux/pm_runtime.h>
+#include <linux/regulator/consumer.h>
+
+#include <media/v4l2-cci.h>
+#include <media/v4l2-ctrls.h>
+#include <media/v4l2-device.h>
+#include <media/v4l2-fwnode.h>
+#include <media/v4l2-mediabus.h>
+#include <media/v4l2-subdev.h>
+
+#define OV64B40_XCLK_FREQ		24000000
+
+#define OV64B40_NATIVE_WIDTH		9286
+#define OV64B40_NATIVE_HEIGHT		6976
+#define OV64B40_PIXEL_ARRAY_TOP		0
+#define OV64B40_PIXEL_ARRAY_LEFT	0
+#define OV64B40_PIXEL_ARRAY_WIDTH	9248
+#define OV64B40_PIXEL_ARRAY_HEIGHT	6944
+
+/* TODO: validate vblank_min, it's not characterized in the datasheet. */
+#define OV64B40_VBLANK_MIN		128
+#define OV64B40_VTS_MAX			0xffffff
+
+#define OV64B40_REG_MEC_LONG_EXPO	CCI_REG24(0x3500)
+#define OV64B40_EXPOSURE_MIN		16
+#define OV64B40_EXPOSURE_MARGIN		32
+
+#define OV64B40_REG_MEC_LONG_GAIN	CCI_REG16(0x3508)
+#define OV64B40_ANA_GAIN_MIN		0x80
+#define OV64B40_ANA_GAIN_MAX		0x7ff
+#define OV64B40_ANA_GAIN_DEFAULT	0x80
+
+#define OV64B40_REG_MEC_LONG_DIG_GAIN	CCI_REG24(0x350a)
+#define OV64B40_DIG_GAIN_MIN		0x400
+#define OV64B40_DIG_GAIN_MAX		0x3fff
+#define OV64B40_DIG_GAIN_DEFAULT	0x400
+
+#define OV64B40_REG_TIMING_CTRL0	CCI_REG16(0x3800)
+#define OV64B40_REG_TIMING_CTRL2	CCI_REG16(0x3802)
+#define OV64B40_REG_TIMING_CTRL4	CCI_REG16(0x3804)
+#define OV64B40_REG_TIMING_CTRL6	CCI_REG16(0x3806)
+#define OV64B40_REG_TIMING_CTRL8	CCI_REG16(0x3808)
+#define OV64B40_REG_TIMING_CTRLA	CCI_REG16(0x380a)
+#define OV64B40_REG_TIMING_CTRLC	CCI_REG16(0x380c)
+#define OV64B40_REG_TIMING_CTRLE	CCI_REG16(0x380e)
+#define OV64B40_REG_TIMING_CTRL10	CCI_REG16(0x3810)
+#define OV64B40_REG_TIMING_CTRL12	CCI_REG16(0x3812)
+
+/*
+ * Careful: a typo in the datasheet calls this register
+ * OV64B40_REG_TIMING_CTRL20.
+ */
+#define OV64B40_REG_TIMING_CTRL14	CCI_REG8(0x3814)
+#define OV64B40_REG_TIMING_CTRL15	CCI_REG8(0x3815)
+#define OV64B40_ODD_INC_SHIFT		4
+#define OV64B40_SKIPPING_CONFIG(_odd, _even) \
+				(((_odd) << OV64B40_ODD_INC_SHIFT) | (_even))
+
+#define OV64B40_REG_TIMING_CTRL_20	CCI_REG8(0x3820)
+#define OV64B40_TIMING_CTRL_20_VFLIP	BIT(2)
+#define OV64B40_TIMING_CTRL_20_VBIN	BIT(1)
+
+#define OV64B40_REG_TIMING_CTRL_21	CCI_REG8(0x3821)
+#define OV64B40_TIMING_CTRL_21_HBIN	BIT(4)
+#define OV64B40_TIMING_CTRL_21_HFLIP	BIT(2)
+#define OV64B40_TIMING_CTRL_21_DSPEED	BIT(0)
+#define OV64B40_TIMING_CTRL_21_HBIN_CONF \
+					(OV64B40_TIMING_CTRL_21_HBIN | \
+					 OV64B40_TIMING_CTRL_21_DSPEED)
+
+#define OV64B40_REG_TIMINGS_VTS_HIGH	CCI_REG8(0x3840)
+#define OV64B40_REG_TIMINGS_VTS_MID	CCI_REG8(0x380e)
+#define OV64B40_REG_TIMINGS_VTS_LOW	CCI_REG8(0x380f)
+
+/* The test pattern control is weirdly named PRE_ISP_2325_D2V2_TOP_1 in TRM. */
+#define OV64B40_REG_TEST_PATTERN	CCI_REG8(0x50c1)
+#define OV64B40_TEST_PATTERN_DISABLED	0x00
+#define OV64B40_TEST_PATTERN_TYPE1	BIT(0)
+#define OV64B40_TEST_PATTERN_TYPE2	(BIT(4) | BIT(0))
+#define OV64B40_TEST_PATTERN_TYPE3	(BIT(5) | BIT(0))
+#define OV64B40_TEST_PATTERN_TYPE4	(BIT(5) | BIT(4) | BIT(0))
+
+#define OV64B40_REG_CHIP_ID		CCI_REG24(0x300a)
+#define OV64B40_CHIP_ID			0x566442
+
+#define OV64B40_REG_SMIA		CCI_REG8(0x0100)
+#define OV64B40_REG_SMIA_STREAMING	BIT(0)
+
+static const char * const ov64b40_supply_names[] = {
+	/* Supplies can be enabled in any order */
+	"avdd",		/* Analog (2.8V) supply */
+	"dovdd",	/* Digital Core (1.8V) supply */
+	"dvdd",		/* IF (1.1V) supply */
+};
+
+static const char * const ov64b40_test_pattern_menu[] = {
+	"Disabled",
+	"Type1",
+	"Type2",
+	"Type3",
+	"Type4",
+};
+
+static const int ov64b40_test_pattern_val[] = {
+	OV64B40_TEST_PATTERN_DISABLED,
+	OV64B40_TEST_PATTERN_TYPE1,
+	OV64B40_TEST_PATTERN_TYPE2,
+	OV64B40_TEST_PATTERN_TYPE3,
+	OV64B40_TEST_PATTERN_TYPE4,
+};
+
+static const unsigned int ov64b40_mbus_codes[] = {
+	MEDIA_BUS_FMT_SBGGR10_1X10,
+	MEDIA_BUS_FMT_SGRBG10_1X10,
+	MEDIA_BUS_FMT_SGBRG10_1X10,
+	MEDIA_BUS_FMT_SRGGB10_1X10,
+};
+
+static const struct cci_reg_sequence ov64b40_init[] = {
+	{ CCI_REG8(0x0103), 0x01 },
+	{ CCI_REG8(0x0102), 0x01 },
+	{ CCI_REG8(0x0102), 0x01 },
+	{ CCI_REG8(0x0301), 0x48 },
+	{ CCI_REG8(0x0303), 0x03 },
+	{ CCI_REG8(0x0304), 0x01 },
+	{ CCI_REG8(0x0305), 0x62 },
+	{ CCI_REG8(0x0306), 0x04 },
+	{ CCI_REG8(0x0307), 0x00 },
+	{ CCI_REG8(0x0309), 0x50 },
+	{ CCI_REG8(0x0316), 0x73 },
+	{ CCI_REG8(0x0320), 0x02 },
+	{ CCI_REG8(0x0321), 0x03 },
+	{ CCI_REG8(0x0323), 0x03 },
+	{ CCI_REG8(0x0324), 0x01 },
+	{ CCI_REG8(0x0325), 0x80 },
+	{ CCI_REG8(0x0326), 0xc3 },
+	{ CCI_REG8(0x0327), 0x03 },
+	{ CCI_REG8(0x0329), 0x00 },
+	{ CCI_REG8(0x032c), 0x00 },
+	{ CCI_REG8(0x032d), 0x01 },
+	{ CCI_REG8(0x032e), 0x06 },
+	{ CCI_REG8(0x032f), 0xa1 },
+	{ CCI_REG8(0x0343), 0x03 },
+	{ CCI_REG8(0x0350), 0x00 },
+	{ CCI_REG8(0x0360), 0x01 },
+	{ CCI_REG8(0x0361), 0x00 },
+	{ CCI_REG8(0x3002), 0x00 },
+	{ CCI_REG8(0x3008), 0x00 },
+	{ CCI_REG8(0x3012), 0x41 },
+	{ CCI_REG8(0x3015), 0x00 },
+	{ CCI_REG8(0x301a), 0xb0 },
+	{ CCI_REG8(0x301e), 0x88 },
+	{ CCI_REG8(0x3025), 0x89 },
+	{ CCI_REG8(0x3218), 0xa1 },
+	{ CCI_REG8(0x3220), 0x12 },
+	{ CCI_REG8(0x3221), 0x28 },
+	{ CCI_REG8(0x3400), 0x04 },
+	{ CCI_REG8(0x3406), 0x10 },
+	{ CCI_REG8(0x340c), 0x00 },
+	{ CCI_REG8(0x340d), 0x00 },
+	{ CCI_REG8(0x340e), 0x60 },
+	{ CCI_REG8(0x3422), 0x00 },
+	{ CCI_REG8(0x3423), 0x00 },
+	{ CCI_REG8(0x3424), 0x15 },
+	{ CCI_REG8(0x3425), 0x40 },
+	{ CCI_REG8(0x3426), 0x10 },
+	{ CCI_REG8(0x3427), 0x10 },
+	{ CCI_REG8(0x3428), 0x00 },
+	{ CCI_REG8(0x3429), 0x00 },
+	{ CCI_REG8(0x3421), 0x08 },
+	{ CCI_REG8(0x3500), 0x00 },
+	{ CCI_REG8(0x3501), 0x0f },
+	{ CCI_REG8(0x3502), 0xec },
+	{ CCI_REG8(0x3504), 0x0c },
+	{ CCI_REG8(0x3508), 0x08 },
+	{ CCI_REG8(0x3509), 0x00 },
+	{ CCI_REG8(0x350a), 0x01 },
+	{ CCI_REG8(0x350b), 0x00 },
+	{ CCI_REG8(0x350c), 0x00 },
+	{ CCI_REG8(0x3540), 0x00 },
+	{ CCI_REG8(0x3541), 0x00 },
+	{ CCI_REG8(0x3542), 0x20 },
+	{ CCI_REG8(0x3544), 0x08 },
+	{ CCI_REG8(0x3548), 0x01 },
+	{ CCI_REG8(0x3549), 0x00 },
+	{ CCI_REG8(0x354a), 0x01 },
+	{ CCI_REG8(0x354b), 0x00 },
+	{ CCI_REG8(0x3580), 0x00 },
+	{ CCI_REG8(0x3581), 0x00 },
+	{ CCI_REG8(0x3582), 0x10 },
+	{ CCI_REG8(0x3584), 0x08 },
+	{ CCI_REG8(0x3588), 0x01 },
+	{ CCI_REG8(0x3589), 0x00 },
+	{ CCI_REG8(0x358a), 0x01 },
+	{ CCI_REG8(0x358b), 0x00 },
+	{ CCI_REG8(0x3605), 0xf9 },
+	{ CCI_REG8(0x3608), 0x8a },
+	{ CCI_REG8(0x360a), 0x5d },
+	{ CCI_REG8(0x360b), 0x50 },
+	{ CCI_REG8(0x360c), 0x92 },
+	{ CCI_REG8(0x360d), 0x05 },
+	{ CCI_REG8(0x360e), 0x08 },
+	{ CCI_REG8(0x3618), 0x80 },
+	{ CCI_REG8(0x361e), 0x80 },
+	{ CCI_REG8(0x3622), 0x0a },
+	{ CCI_REG8(0x3623), 0x08 },
+	{ CCI_REG8(0x3628), 0x99 },
+	{ CCI_REG8(0x362b), 0x77 },
+	{ CCI_REG8(0x362d), 0x0a },
+	{ CCI_REG8(0x3659), 0x48 },
+	{ CCI_REG8(0x365a), 0x2b },
+	{ CCI_REG8(0x365b), 0x0a },
+	{ CCI_REG8(0x3684), 0x8b },
+	{ CCI_REG8(0x3685), 0x00 },
+	{ CCI_REG8(0x3688), 0x02 },
+	{ CCI_REG8(0x3689), 0x88 },
+	{ CCI_REG8(0x368a), 0x26 },
+	{ CCI_REG8(0x368d), 0x00 },
+	{ CCI_REG8(0x368e), 0x78 },
+	{ CCI_REG8(0x368f), 0x00 },
+	{ CCI_REG8(0x3694), 0x7f },
+	{ CCI_REG8(0x3695), 0x00 },
+	{ CCI_REG8(0x3696), 0x11 },
+	{ CCI_REG8(0x3699), 0x19 },
+	{ CCI_REG8(0x369a), 0x00 },
+	{ CCI_REG8(0x369f), 0x20 },
+	{ CCI_REG8(0x36a4), 0x00 },
+	{ CCI_REG8(0x36a6), 0x2b },
+	{ CCI_REG8(0x36a7), 0x19 },
+	{ CCI_REG8(0x36a8), 0x1f },
+	{ CCI_REG8(0x36aa), 0x12 },
+	{ CCI_REG8(0x36ab), 0x28 },
+	{ CCI_REG8(0x36b0), 0x12 },
+	{ CCI_REG8(0x36b1), 0x28 },
+	{ CCI_REG8(0x36b2), 0x12 },
+	{ CCI_REG8(0x36b3), 0x28 },
+	{ CCI_REG8(0x36b4), 0x12 },
+	{ CCI_REG8(0x36b5), 0x28 },
+	{ CCI_REG8(0x36b6), 0x12 },
+	{ CCI_REG8(0x36b7), 0x28 },
+	{ CCI_REG8(0x36b8), 0x12 },
+	{ CCI_REG8(0x36b9), 0x28 },
+	{ CCI_REG8(0x36ba), 0x12 },
+	{ CCI_REG8(0x36bb), 0x28 },
+	{ CCI_REG8(0x36bc), 0x12 },
+	{ CCI_REG8(0x36bd), 0x28 },
+	{ CCI_REG8(0x36be), 0x12 },
+	{ CCI_REG8(0x36bf), 0x28 },
+	{ CCI_REG8(0x36c0), 0x12 },
+	{ CCI_REG8(0x36c1), 0x28 },
+	{ CCI_REG8(0x3700), 0x2c },
+	{ CCI_REG8(0x3701), 0x40 },
+	{ CCI_REG8(0x3702), 0x4b },
+	{ CCI_REG8(0x3703), 0x40 },
+	{ CCI_REG8(0x3706), 0x3a },
+	{ CCI_REG8(0x3708), 0x59 },
+	{ CCI_REG8(0x3709), 0xf8 },
+	{ CCI_REG8(0x370b), 0x82 },
+	{ CCI_REG8(0x3711), 0x00 },
+	{ CCI_REG8(0x3712), 0x50 },
+	{ CCI_REG8(0x3713), 0x00 },
+	{ CCI_REG8(0x3714), 0x67 },
+	{ CCI_REG8(0x3716), 0x40 },
+	{ CCI_REG8(0x3717), 0x82 },
+	{ CCI_REG8(0x371d), 0x02 },
+	{ CCI_REG8(0x371e), 0x12 },
+	{ CCI_REG8(0x371f), 0x02 },
+	{ CCI_REG8(0x3720), 0x08 },
+	{ CCI_REG8(0x3721), 0x02 },
+	{ CCI_REG8(0x3723), 0x00 },
+	{ CCI_REG8(0x3724), 0x0d },
+	{ CCI_REG8(0x3725), 0x22 },
+	{ CCI_REG8(0x372d), 0x00 },
+	{ CCI_REG8(0x3730), 0x14 },
+	{ CCI_REG8(0x3731), 0x14 },
+	{ CCI_REG8(0x3732), 0x07 },
+	{ CCI_REG8(0x3734), 0x59 },
+	{ CCI_REG8(0x3736), 0x0c },
+	{ CCI_REG8(0x3738), 0x01 },
+	{ CCI_REG8(0x3739), 0x18 },
+	{ CCI_REG8(0x373b), 0x05 },
+	{ CCI_REG8(0x3741), 0x48 },
+	{ CCI_REG8(0x3743), 0xef },
+	{ CCI_REG8(0x3744), 0x0f },
+	{ CCI_REG8(0x3745), 0xff },
+	{ CCI_REG8(0x3746), 0x00 },
+	{ CCI_REG8(0x3748), 0x01 },
+	{ CCI_REG8(0x3749), 0x18 },
+	{ CCI_REG8(0x3751), 0x49 },
+	{ CCI_REG8(0x3753), 0xff },
+	{ CCI_REG8(0x3754), 0x00 },
+	{ CCI_REG8(0x3755), 0x40 },
+	{ CCI_REG8(0x3758), 0x88 },
+	{ CCI_REG8(0x3759), 0x88 },
+	{ CCI_REG8(0x375a), 0x88 },
+	{ CCI_REG8(0x375b), 0x88 },
+	{ CCI_REG8(0x375c), 0x88 },
+	{ CCI_REG8(0x375e), 0x00 },
+	{ CCI_REG8(0x375f), 0x02 },
+	{ CCI_REG8(0x3760), 0x08 },
+	{ CCI_REG8(0x3761), 0x10 },
+	{ CCI_REG8(0x3762), 0x08 },
+	{ CCI_REG8(0x3763), 0x08 },
+	{ CCI_REG8(0x3764), 0x08 },
+	{ CCI_REG8(0x3765), 0x10 },
+	{ CCI_REG8(0x3766), 0x18 },
+	{ CCI_REG8(0x3767), 0x10 },
+	{ CCI_REG8(0x3768), 0x00 },
+	{ CCI_REG8(0x3769), 0x08 },
+	{ CCI_REG8(0x376a), 0x10 },
+	{ CCI_REG8(0x376b), 0x00 },
+	{ CCI_REG8(0x376c), 0x28 },
+	{ CCI_REG8(0x3770), 0x19 },
+	{ CCI_REG8(0x3771), 0x01 },
+	{ CCI_REG8(0x3772), 0x01 },
+	{ CCI_REG8(0x3773), 0x01 },
+	{ CCI_REG8(0x3774), 0x1a },
+	{ CCI_REG8(0x3775), 0x01 },
+	{ CCI_REG8(0x3776), 0x01 },
+	{ CCI_REG8(0x3777), 0x01 },
+	{ CCI_REG8(0x3780), 0xc8 },
+	{ CCI_REG8(0x3791), 0x38 },
+	{ CCI_REG8(0x3793), 0x3a },
+	{ CCI_REG8(0x3795), 0x3a },
+	{ CCI_REG8(0x3797), 0x80 },
+	{ CCI_REG8(0x379c), 0x14 },
+	{ CCI_REG8(0x379d), 0x14 },
+	{ CCI_REG8(0x3799), 0x82 },
+	{ CCI_REG8(0x379b), 0x82 },
+	{ CCI_REG8(0x37a0), 0x07 },
+	{ CCI_REG8(0x37a1), 0x99 },
+	{ CCI_REG8(0x37a2), 0x11 },
+	{ CCI_REG8(0x37af), 0x00 },
+	{ CCI_REG8(0x37f7), 0x00 },
+	{ CCI_REG8(0x37f8), 0x5a },
+	{ CCI_REG8(0x37ff), 0x18 },
+	{ CCI_REG8(0x3800), 0x00 },
+	{ CCI_REG8(0x3801), 0x00 },
+	{ CCI_REG8(0x3802), 0x00 },
+	{ CCI_REG8(0x3803), 0x00 },
+	{ CCI_REG8(0x3804), 0x24 },
+	{ CCI_REG8(0x3805), 0x3f },
+	{ CCI_REG8(0x3806), 0x1b },
+	{ CCI_REG8(0x3807), 0x3f },
+	{ CCI_REG8(0x3808), 0x12 },
+	{ CCI_REG8(0x3809), 0x10 },
+	{ CCI_REG8(0x380a), 0x0d },
+	{ CCI_REG8(0x380b), 0x90 },
+	{ CCI_REG8(0x380c), 0x03 },
+	{ CCI_REG8(0x380d), 0xa8 },
+	{ CCI_REG8(0x380e), 0x10 },
+	{ CCI_REG8(0x380f), 0x0c },
+	{ CCI_REG8(0x3810), 0x00 },
+	{ CCI_REG8(0x3811), 0x08 },
+	{ CCI_REG8(0x3812), 0x00 },
+	{ CCI_REG8(0x3813), 0x09 },
+	{ CCI_REG8(0x3814), 0x11 },
+	{ CCI_REG8(0x3815), 0x11 },
+	{ CCI_REG8(0x3820), 0x42 },
+	{ CCI_REG8(0x3821), 0x04 },
+	{ CCI_REG8(0x3822), 0x10 },
+	{ CCI_REG8(0x3823), 0x04 },
+	{ CCI_REG8(0x3828), 0x0f },
+	{ CCI_REG8(0x382a), 0x80 },
+	{ CCI_REG8(0x382e), 0x41 },
+	{ CCI_REG8(0x3830), 0x0b },
+	{ CCI_REG8(0x3837), 0x0c },
+	{ CCI_REG8(0x383e), 0x00 },
+	{ CCI_REG8(0x383f), 0x0c },
+	{ CCI_REG8(0x3840), 0x00 },
+	{ CCI_REG8(0x3847), 0x00 },
+	{ CCI_REG8(0x384a), 0x00 },
+	{ CCI_REG8(0x384c), 0x03 },
+	{ CCI_REG8(0x384d), 0xa8 },
+	{ CCI_REG8(0x3856), 0x00 },
+	{ CCI_REG8(0x3858), 0x00 },
+	{ CCI_REG8(0x3859), 0x00 },
+	{ CCI_REG8(0x3888), 0x00 },
+	{ CCI_REG8(0x3889), 0x60 },
+	{ CCI_REG8(0x388a), 0x00 },
+	{ CCI_REG8(0x388b), 0x18 },
+	{ CCI_REG8(0x388c), 0x23 },
+	{ CCI_REG8(0x388d), 0x80 },
+	{ CCI_REG8(0x388e), 0x0d },
+	{ CCI_REG8(0x388f), 0x70 },
+	{ CCI_REG8(0x3894), 0x00 },
+	{ CCI_REG8(0x3900), 0x0c },
+	{ CCI_REG8(0x3901), 0x06 },
+	{ CCI_REG8(0x3902), 0xdf },
+	{ CCI_REG8(0x3903), 0xfb },
+	{ CCI_REG8(0x3904), 0xff },
+	{ CCI_REG8(0x3906), 0x00 },
+	{ CCI_REG8(0x3907), 0x00 },
+	{ CCI_REG8(0x3908), 0x57 },
+	{ CCI_REG8(0x3909), 0xde },
+	{ CCI_REG8(0x390c), 0x3f },
+	{ CCI_REG8(0x390e), 0x3d },
+	{ CCI_REG8(0x390f), 0x54 },
+	{ CCI_REG8(0x3911), 0x3c },
+	{ CCI_REG8(0x3912), 0x03 },
+	{ CCI_REG8(0x3913), 0x41 },
+	{ CCI_REG8(0x3914), 0x40 },
+	{ CCI_REG8(0x3919), 0x26 },
+	{ CCI_REG8(0x391b), 0xa2 },
+	{ CCI_REG8(0x391c), 0x69 },
+	{ CCI_REG8(0x3920), 0x88 },
+	{ CCI_REG8(0x3921), 0x00 },
+	{ CCI_REG8(0x3924), 0x84 },
+	{ CCI_REG8(0x3925), 0x0f },
+	{ CCI_REG8(0x3926), 0xfe },
+	{ CCI_REG8(0x3930), 0x03 },
+	{ CCI_REG8(0x3932), 0x03 },
+	{ CCI_REG8(0x3933), 0x01 },
+	{ CCI_REG8(0x3935), 0x01 },
+	{ CCI_REG8(0x393a), 0x03 },
+	{ CCI_REG8(0x393c), 0x03 },
+	{ CCI_REG8(0x393d), 0x01 },
+	{ CCI_REG8(0x393f), 0x01 },
+	{ CCI_REG8(0x3945), 0x00 },
+	{ CCI_REG8(0x3946), 0x00 },
+	{ CCI_REG8(0x3949), 0x01 },
+	{ CCI_REG8(0x394b), 0x00 },
+	{ CCI_REG8(0x394c), 0x16 },
+	{ CCI_REG8(0x3953), 0x0c },
+	{ CCI_REG8(0x3954), 0x0c },
+	{ CCI_REG8(0x3955), 0x00 },
+	{ CCI_REG8(0x3956), 0x00 },
+	{ CCI_REG8(0x3968), 0x00 },
+	{ CCI_REG8(0x396a), 0x00 },
+	{ CCI_REG8(0x3973), 0x02 },
+	{ CCI_REG8(0x3977), 0x04 },
+	{ CCI_REG8(0x3978), 0x02 },
+	{ CCI_REG8(0x3979), 0x04 },
+	{ CCI_REG8(0x397d), 0x04 },
+	{ CCI_REG8(0x3981), 0x04 },
+	{ CCI_REG8(0x3983), 0x04 },
+	{ CCI_REG8(0x3987), 0x04 },
+	{ CCI_REG8(0x3989), 0x04 },
+	{ CCI_REG8(0x3997), 0x00 },
+	{ CCI_REG8(0x399b), 0x4b },
+	{ CCI_REG8(0x399c), 0x00 },
+	{ CCI_REG8(0x39a2), 0x02 },
+	{ CCI_REG8(0x39af), 0x01 },
+	{ CCI_REG8(0x39b7), 0x01 },
+	{ CCI_REG8(0x39bf), 0x01 },
+	{ CCI_REG8(0x39cb), 0x00 },
+	{ CCI_REG8(0x39cd), 0x18 },
+	{ CCI_REG8(0x39ce), 0x48 },
+	{ CCI_REG8(0x39cf), 0x48 },
+	{ CCI_REG8(0x39d0), 0x48 },
+	{ CCI_REG8(0x39d2), 0x1b },
+	{ CCI_REG8(0x39d3), 0x4b },
+	{ CCI_REG8(0x39d4), 0x4b },
+	{ CCI_REG8(0x39d5), 0x4b },
+	{ CCI_REG8(0x39d7), 0x01 },
+	{ CCI_REG8(0x39dc), 0x00 },
+	{ CCI_REG8(0x39e2), 0x20 },
+	{ CCI_REG8(0x39e4), 0x20 },
+	{ CCI_REG8(0x39e6), 0x20 },
+	{ CCI_REG8(0x39e8), 0x20 },
+	{ CCI_REG8(0x39eb), 0x01 },
+	{ CCI_REG8(0x39fd), 0x01 },
+	{ CCI_REG8(0x39f5), 0x04 },
+	{ CCI_REG8(0x39f6), 0x1a },
+	{ CCI_REG8(0x39f7), 0x00 },
+	{ CCI_REG8(0x39f8), 0x00 },
+	{ CCI_REG8(0x39f9), 0x00 },
+	{ CCI_REG8(0x39ff), 0x00 },
+	{ CCI_REG8(0x3a01), 0x03 },
+	{ CCI_REG8(0x3a03), 0x00 },
+	{ CCI_REG8(0x3a05), 0x05 },
+	{ CCI_REG8(0x3a07), 0x07 },
+	{ CCI_REG8(0x3a17), 0x04 },
+	{ CCI_REG8(0x3a2d), 0x00 },
+	{ CCI_REG8(0x3a2e), 0x00 },
+	{ CCI_REG8(0x3a2f), 0x20 },
+	{ CCI_REG8(0x3a30), 0x20 },
+	{ CCI_REG8(0x3a31), 0x20 },
+	{ CCI_REG8(0x3a32), 0x20 },
+	{ CCI_REG8(0x3a33), 0x20 },
+	{ CCI_REG8(0x3a34), 0x20 },
+	{ CCI_REG8(0x3a35), 0x55 },
+	{ CCI_REG8(0x3a43), 0x1e },
+	{ CCI_REG8(0x3a44), 0x0f },
+	{ CCI_REG8(0x3a45), 0x07 },
+	{ CCI_REG8(0x3a46), 0x03 },
+	{ CCI_REG8(0x3a4b), 0x08 },
+	{ CCI_REG8(0x3a4c), 0x08 },
+	{ CCI_REG8(0x3a4d), 0x08 },
+	{ CCI_REG8(0x3a4e), 0x08 },
+	{ CCI_REG8(0x3a4f), 0x08 },
+	{ CCI_REG8(0x3a50), 0x08 },
+	{ CCI_REG8(0x3a51), 0x08 },
+	{ CCI_REG8(0x3a52), 0x08 },
+	{ CCI_REG8(0x3a53), 0x88 },
+	{ CCI_REG8(0x3a54), 0x88 },
+	{ CCI_REG8(0x3a55), 0x88 },
+	{ CCI_REG8(0x3a56), 0x88 },
+	{ CCI_REG8(0x3a57), 0x15 },
+	{ CCI_REG8(0x3a58), 0x00 },
+	{ CCI_REG8(0x3a59), 0x00 },
+	{ CCI_REG8(0x3a5a), 0x01 },
+	{ CCI_REG8(0x3a5f), 0x3f },
+	{ CCI_REG8(0x3a60), 0x3f },
+	{ CCI_REG8(0x3a61), 0x3f },
+	{ CCI_REG8(0x3a62), 0x3f },
+	{ CCI_REG8(0x3a63), 0x3f },
+	{ CCI_REG8(0x3a64), 0x3f },
+	{ CCI_REG8(0x3a65), 0x3f },
+	{ CCI_REG8(0x3a66), 0x3f },
+	{ CCI_REG8(0x3a67), 0x1f },
+	{ CCI_REG8(0x3a68), 0x1f },
+	{ CCI_REG8(0x3a69), 0x1f },
+	{ CCI_REG8(0x3a6a), 0x1f },
+	{ CCI_REG8(0x3a6b), 0x1f },
+	{ CCI_REG8(0x3a6c), 0x1f },
+	{ CCI_REG8(0x3a6d), 0x1f },
+	{ CCI_REG8(0x3a6e), 0x1f },
+	{ CCI_REG8(0x3a73), 0x13 },
+	{ CCI_REG8(0x3a74), 0x10 },
+	{ CCI_REG8(0x3a75), 0x0b },
+	{ CCI_REG8(0x3a76), 0x00 },
+	{ CCI_REG8(0x3a77), 0x13 },
+	{ CCI_REG8(0x3a78), 0x10 },
+	{ CCI_REG8(0x3a79), 0x0b },
+	{ CCI_REG8(0x3a7a), 0x00 },
+	{ CCI_REG8(0x3a7d), 0x60 },
+	{ CCI_REG8(0x3d85), 0x0b },
+	{ CCI_REG8(0x3d86), 0x12 },
+	{ CCI_REG8(0x3d87), 0x28 },
+	{ CCI_REG8(0x3d8c), 0x07 },
+	{ CCI_REG8(0x3d8d), 0xb0 },
+	{ CCI_REG8(0x3daa), 0x70 },
+	{ CCI_REG8(0x3dab), 0x10 },
+	{ CCI_REG8(0x3dac), 0x76 },
+	{ CCI_REG8(0x3dad), 0xa0 },
+	{ CCI_REG8(0x3dae), 0x77 },
+	{ CCI_REG8(0x3daf), 0x1f },
+	{ CCI_REG8(0x3f00), 0x10 },
+	{ CCI_REG8(0x4009), 0x02 },
+	{ CCI_REG8(0x400e), 0xc6 },
+	{ CCI_REG8(0x400f), 0x00 },
+	{ CCI_REG8(0x4010), 0xe8 },
+	{ CCI_REG8(0x4011), 0x01 },
+	{ CCI_REG8(0x4012), 0x0c },
+	{ CCI_REG8(0x4015), 0x00 },
+	{ CCI_REG8(0x4016), 0x0f },
+	{ CCI_REG8(0x4017), 0x00 },
+	{ CCI_REG8(0x4018), 0x07 },
+	{ CCI_REG8(0x401a), 0x40 },
+	{ CCI_REG8(0x401b), 0x04 },
+	{ CCI_REG8(0x401e), 0x01 },
+	{ CCI_REG8(0x401f), 0x00 },
+	{ CCI_REG8(0x4020), 0x04 },
+	{ CCI_REG8(0x4021), 0x00 },
+	{ CCI_REG8(0x4022), 0x04 },
+	{ CCI_REG8(0x4023), 0x00 },
+	{ CCI_REG8(0x4024), 0x04 },
+	{ CCI_REG8(0x4025), 0x00 },
+	{ CCI_REG8(0x4026), 0x04 },
+	{ CCI_REG8(0x4027), 0x00 },
+	{ CCI_REG8(0x4028), 0x01 },
+	{ CCI_REG8(0x4029), 0x08 },
+	{ CCI_REG8(0x40c2), 0x12 },
+	{ CCI_REG8(0x40c3), 0x28 },
+	{ CCI_REG8(0x4288), 0xa7 },
+	{ CCI_REG8(0x4289), 0x20 },
+	{ CCI_REG8(0x428c), 0x00 },
+	{ CCI_REG8(0x428d), 0x80 },
+	{ CCI_REG8(0x428e), 0x04 },
+	{ CCI_REG8(0x4500), 0x08 },
+	{ CCI_REG8(0x4502), 0x00 },
+	{ CCI_REG8(0x4504), 0x80 },
+	{ CCI_REG8(0x4506), 0x01 },
+	{ CCI_REG8(0x4509), 0x05 },
+	{ CCI_REG8(0x450c), 0x00 },
+	{ CCI_REG8(0x450d), 0x20 },
+	{ CCI_REG8(0x450e), 0x00 },
+	{ CCI_REG8(0x450f), 0x00 },
+	{ CCI_REG8(0x4510), 0x00 },
+	{ CCI_REG8(0x4523), 0x00 },
+	{ CCI_REG8(0x4526), 0x00 },
+	{ CCI_REG8(0x4540), 0x12 },
+	{ CCI_REG8(0x4541), 0x28 },
+	{ CCI_REG8(0x4542), 0x00 },
+	{ CCI_REG8(0x4543), 0x00 },
+	{ CCI_REG8(0x4544), 0x00 },
+	{ CCI_REG8(0x4545), 0x00 },
+	{ CCI_REG8(0x4546), 0x00 },
+	{ CCI_REG8(0x4547), 0x10 },
+	{ CCI_REG8(0x4580), 0x01 },
+	{ CCI_REG8(0x4589), 0x12 },
+	{ CCI_REG8(0x458a), 0x28 },
+	{ CCI_REG8(0x45c0), 0x61 },
+	{ CCI_REG8(0x45c9), 0x12 },
+	{ CCI_REG8(0x45ca), 0x28 },
+	{ CCI_REG8(0x45cb), 0x30 },
+	{ CCI_REG8(0x4602), 0x00 },
+	{ CCI_REG8(0x4603), 0x15 },
+	{ CCI_REG8(0x4606), 0x12 },
+	{ CCI_REG8(0x4607), 0x28 },
+	{ CCI_REG8(0x4609), 0x20 },
+	{ CCI_REG8(0x460b), 0x03 },
+	{ CCI_REG8(0x4640), 0x00 },
+	{ CCI_REG8(0x4641), 0x47 },
+	{ CCI_REG8(0x4643), 0x0c },
+	{ CCI_REG8(0x4644), 0x40 },
+	{ CCI_REG8(0x4645), 0xb3 },
+	{ CCI_REG8(0x4648), 0x12 },
+	{ CCI_REG8(0x4649), 0x28 },
+	{ CCI_REG8(0x464c), 0x01 },
+	{ CCI_REG8(0x464a), 0x00 },
+	{ CCI_REG8(0x464b), 0x30 },
+	{ CCI_REG8(0x4800), 0x64 },
+	{ CCI_REG8(0x4802), 0x00 },
+	{ CCI_REG8(0x480b), 0x10 },
+	{ CCI_REG8(0x480c), 0x80 },
+	{ CCI_REG8(0x480e), 0x04 },
+	{ CCI_REG8(0x480f), 0x32 },
+	{ CCI_REG8(0x481b), 0x3c },
+	{ CCI_REG8(0x481f), 0x30 },
+	{ CCI_REG8(0x4833), 0x20 },
+	{ CCI_REG8(0x4837), 0x09 },
+	{ CCI_REG8(0x484b), 0x27 },
+	{ CCI_REG8(0x4850), 0x47 },
+	{ CCI_REG8(0x4853), 0x04 },
+	{ CCI_REG8(0x4854), 0x08 },
+	{ CCI_REG8(0x4860), 0x00 },
+	{ CCI_REG8(0x4861), 0xec },
+	{ CCI_REG8(0x4862), 0x05 },
+	{ CCI_REG8(0x4883), 0x05 },
+	{ CCI_REG8(0x4885), 0x3c },
+	{ CCI_REG8(0x4888), 0x10 },
+	{ CCI_REG8(0x4889), 0x03 },
+	{ CCI_REG8(0x488a), 0x10 },
+	{ CCI_REG8(0x4909), 0x02 },
+	{ CCI_REG8(0x4910), 0xe8 },
+	{ CCI_REG8(0x4911), 0x01 },
+	{ CCI_REG8(0x4912), 0x0c },
+	{ CCI_REG8(0x4915), 0x00 },
+	{ CCI_REG8(0x4916), 0x0f },
+	{ CCI_REG8(0x4917), 0x00 },
+	{ CCI_REG8(0x4918), 0x07 },
+	{ CCI_REG8(0x491a), 0x40 },
+	{ CCI_REG8(0x491b), 0x04 },
+	{ CCI_REG8(0x491e), 0x01 },
+	{ CCI_REG8(0x491f), 0x00 },
+	{ CCI_REG8(0x4920), 0x04 },
+	{ CCI_REG8(0x4921), 0x00 },
+	{ CCI_REG8(0x4922), 0x04 },
+	{ CCI_REG8(0x4923), 0x00 },
+	{ CCI_REG8(0x4924), 0x04 },
+	{ CCI_REG8(0x4925), 0x00 },
+	{ CCI_REG8(0x4926), 0x04 },
+	{ CCI_REG8(0x4927), 0x00 },
+	{ CCI_REG8(0x4929), 0x08 },
+	{ CCI_REG8(0x49c2), 0x12 },
+	{ CCI_REG8(0x49c3), 0x28 },
+	{ CCI_REG8(0x4a09), 0x02 },
+	{ CCI_REG8(0x4a10), 0xe8 },
+	{ CCI_REG8(0x4a11), 0x01 },
+	{ CCI_REG8(0x4a12), 0x0c },
+	{ CCI_REG8(0x4a15), 0x00 },
+	{ CCI_REG8(0x4a16), 0x0f },
+	{ CCI_REG8(0x4a17), 0x00 },
+	{ CCI_REG8(0x4a18), 0x07 },
+	{ CCI_REG8(0x4a1a), 0x40 },
+	{ CCI_REG8(0x4a1b), 0x04 },
+	{ CCI_REG8(0x4a1e), 0x01 },
+	{ CCI_REG8(0x4a1f), 0x00 },
+	{ CCI_REG8(0x4a20), 0x04 },
+	{ CCI_REG8(0x4a21), 0x00 },
+	{ CCI_REG8(0x4a22), 0x04 },
+	{ CCI_REG8(0x4a23), 0x00 },
+	{ CCI_REG8(0x4a24), 0x04 },
+	{ CCI_REG8(0x4a25), 0x00 },
+	{ CCI_REG8(0x4a26), 0x04 },
+	{ CCI_REG8(0x4a27), 0x00 },
+	{ CCI_REG8(0x4a29), 0x08 },
+	{ CCI_REG8(0x4ac2), 0x12 },
+	{ CCI_REG8(0x4ac3), 0x28 },
+	{ CCI_REG8(0x4d00), 0x04 },
+	{ CCI_REG8(0x4d01), 0xf5 },
+	{ CCI_REG8(0x4d02), 0xb8 },
+	{ CCI_REG8(0x4d03), 0x76 },
+	{ CCI_REG8(0x4d04), 0x04 },
+	{ CCI_REG8(0x4d05), 0xcc },
+	{ CCI_REG8(0x5000), 0xf7 },
+	{ CCI_REG8(0x5001), 0x01 },
+	{ CCI_REG8(0x5002), 0xf7 },
+	{ CCI_REG8(0x5003), 0x00 },
+	{ CCI_REG8(0x5004), 0x80 },
+	{ CCI_REG8(0x5005), 0x42 },
+	{ CCI_REG8(0x5006), 0x04 },
+	{ CCI_REG8(0x5007), 0x02 },
+	{ CCI_REG8(0x5015), 0x20 },
+	{ CCI_REG8(0x5053), 0x04 },
+	{ CCI_REG8(0x5060), 0x00 },
+	{ CCI_REG8(0x5061), 0x30 },
+	{ CCI_REG8(0x5062), 0x00 },
+	{ CCI_REG8(0x5063), 0x30 },
+	{ CCI_REG8(0x5064), 0x23 },
+	{ CCI_REG8(0x5065), 0xe0 },
+	{ CCI_REG8(0x5066), 0x1a },
+	{ CCI_REG8(0x5067), 0xe0 },
+	{ CCI_REG8(0x5068), 0x02 },
+	{ CCI_REG8(0x5069), 0x10 },
+	{ CCI_REG8(0x506a), 0x10 },
+	{ CCI_REG8(0x506b), 0x04 },
+	{ CCI_REG8(0x506c), 0x04 },
+	{ CCI_REG8(0x506d), 0x0c },
+	{ CCI_REG8(0x506e), 0x0c },
+	{ CCI_REG8(0x506f), 0x04 },
+	{ CCI_REG8(0x5070), 0x04 },
+	{ CCI_REG8(0x5071), 0x0c },
+	{ CCI_REG8(0x5072), 0x0c },
+	{ CCI_REG8(0x5073), 0x01 },
+	{ CCI_REG8(0x5074), 0x01 },
+	{ CCI_REG8(0x5075), 0xaa },
+	{ CCI_REG8(0x5083), 0x00 },
+	{ CCI_REG8(0x50c1), 0x00 },
+	{ CCI_REG8(0x5110), 0x70 },
+	{ CCI_REG8(0x5111), 0x10 },
+	{ CCI_REG8(0x5112), 0x77 },
+	{ CCI_REG8(0x5113), 0x1f },
+	{ CCI_REG8(0x5114), 0x01 },
+	{ CCI_REG8(0x5180), 0xc1 },
+	{ CCI_REG8(0x518a), 0x04 },
+	{ CCI_REG8(0x51b0), 0x00 },
+	{ CCI_REG8(0x51d0), 0xc2 },
+	{ CCI_REG8(0x51d1), 0x68 },
+	{ CCI_REG8(0x51d2), 0xff },
+	{ CCI_REG8(0x51d3), 0x1c },
+	{ CCI_REG8(0x51d8), 0x08 },
+	{ CCI_REG8(0x51d9), 0x10 },
+	{ CCI_REG8(0x51da), 0x02 },
+	{ CCI_REG8(0x51db), 0x02 },
+	{ CCI_REG8(0x51dc), 0x06 },
+	{ CCI_REG8(0x51dd), 0x06 },
+	{ CCI_REG8(0x51de), 0x02 },
+	{ CCI_REG8(0x51df), 0x06 },
+	{ CCI_REG8(0x51e0), 0x0a },
+	{ CCI_REG8(0x51e1), 0x0e },
+	{ CCI_REG8(0x51e2), 0x00 },
+	{ CCI_REG8(0x51e3), 0x00 },
+	{ CCI_REG8(0x51ee), 0x01 },
+	{ CCI_REG8(0x51ef), 0x02 },
+	{ CCI_REG8(0x51f0), 0x04 },
+	{ CCI_REG8(0x51f1), 0xaa },
+	{ CCI_REG8(0x5200), 0x12 },
+	{ CCI_REG8(0x5201), 0x20 },
+	{ CCI_REG8(0x5202), 0x0d },
+	{ CCI_REG8(0x5203), 0xa0 },
+	{ CCI_REG8(0x5205), 0x10 },
+	{ CCI_REG8(0x5207), 0x10 },
+	{ CCI_REG8(0x5208), 0x12 },
+	{ CCI_REG8(0x5209), 0x00 },
+	{ CCI_REG8(0x520a), 0x0d },
+	{ CCI_REG8(0x520b), 0x80 },
+	{ CCI_REG8(0x520d), 0x08 },
+	{ CCI_REG8(0x5250), 0x06 },
+	{ CCI_REG8(0x5251), 0x00 },
+	{ CCI_REG8(0x525b), 0x00 },
+	{ CCI_REG8(0x525d), 0x00 },
+	{ CCI_REG8(0x527a), 0x00 },
+	{ CCI_REG8(0x527b), 0x38 },
+	{ CCI_REG8(0x527c), 0x00 },
+	{ CCI_REG8(0x527d), 0x4b },
+	{ CCI_REG8(0x5286), 0x1b },
+	{ CCI_REG8(0x5287), 0x40 },
+	{ CCI_REG8(0x5298), 0x00 },
+	{ CCI_REG8(0x5299), 0x50 },
+	{ CCI_REG8(0x529a), 0x00 },
+	{ CCI_REG8(0x529b), 0x50 },
+	{ CCI_REG8(0x529c), 0x00 },
+	{ CCI_REG8(0x529d), 0x50 },
+	{ CCI_REG8(0x529e), 0x00 },
+	{ CCI_REG8(0x529f), 0x50 },
+	{ CCI_REG8(0x52a0), 0x00 },
+	{ CCI_REG8(0x52a1), 0x50 },
+	{ CCI_REG8(0x52a2), 0x00 },
+	{ CCI_REG8(0x52a3), 0x50 },
+	{ CCI_REG8(0x52a4), 0x00 },
+	{ CCI_REG8(0x52a5), 0x50 },
+	{ CCI_REG8(0x52a6), 0x00 },
+	{ CCI_REG8(0x52a7), 0x50 },
+	{ CCI_REG8(0x52a8), 0x00 },
+	{ CCI_REG8(0x52a9), 0x50 },
+	{ CCI_REG8(0x52aa), 0x00 },
+	{ CCI_REG8(0x52ab), 0x50 },
+	{ CCI_REG8(0x52ac), 0x00 },
+	{ CCI_REG8(0x52ad), 0x50 },
+	{ CCI_REG8(0x52ae), 0x00 },
+	{ CCI_REG8(0x52af), 0x50 },
+	{ CCI_REG8(0x52d0), 0x20 },
+	{ CCI_REG8(0x52d1), 0x20 },
+	{ CCI_REG8(0x52d2), 0x20 },
+	{ CCI_REG8(0x52d3), 0x20 },
+	{ CCI_REG8(0x52d4), 0x20 },
+	{ CCI_REG8(0x52d5), 0x20 },
+	{ CCI_REG8(0x52d6), 0x20 },
+	{ CCI_REG8(0x52d7), 0x20 },
+	{ CCI_REG8(0x52d8), 0x20 },
+	{ CCI_REG8(0x52d9), 0x20 },
+	{ CCI_REG8(0x52da), 0x20 },
+	{ CCI_REG8(0x52db), 0x20 },
+	{ CCI_REG8(0x52dc), 0x20 },
+	{ CCI_REG8(0x52dd), 0x20 },
+	{ CCI_REG8(0x52de), 0x20 },
+	{ CCI_REG8(0x52df), 0x20 },
+	{ CCI_REG8(0x52f0), 0x08 },
+	{ CCI_REG8(0x52f1), 0x07 },
+	{ CCI_REG8(0x52f2), 0x06 },
+	{ CCI_REG8(0x52f3), 0x05 },
+	{ CCI_REG8(0x52f4), 0x0a },
+	{ CCI_REG8(0x52f5), 0x02 },
+	{ CCI_REG8(0x52f6), 0x01 },
+	{ CCI_REG8(0x52f7), 0x09 },
+	{ CCI_REG8(0x52f8), 0x0c },
+	{ CCI_REG8(0x52f9), 0x04 },
+	{ CCI_REG8(0x52fa), 0x03 },
+	{ CCI_REG8(0x52fb), 0x0b },
+	{ CCI_REG8(0x52fc), 0x10 },
+	{ CCI_REG8(0x52fd), 0x0f },
+	{ CCI_REG8(0x52fe), 0x0e },
+	{ CCI_REG8(0x52ff), 0x0d },
+	{ CCI_REG8(0x5300), 0x00 },
+	{ CCI_REG8(0x5301), 0x00 },
+	{ CCI_REG8(0x5302), 0x00 },
+	{ CCI_REG8(0x5303), 0x00 },
+	{ CCI_REG8(0x5304), 0x00 },
+	{ CCI_REG8(0x5305), 0x00 },
+	{ CCI_REG8(0x5306), 0x00 },
+	{ CCI_REG8(0x5307), 0x00 },
+	{ CCI_REG8(0x5308), 0x00 },
+	{ CCI_REG8(0x5309), 0x00 },
+	{ CCI_REG8(0x530a), 0x00 },
+	{ CCI_REG8(0x530b), 0x00 },
+	{ CCI_REG8(0x530c), 0x00 },
+	{ CCI_REG8(0x530d), 0x00 },
+	{ CCI_REG8(0x530e), 0x00 },
+	{ CCI_REG8(0x530f), 0x00 },
+	{ CCI_REG8(0x5310), 0x03 },
+	{ CCI_REG8(0x5311), 0xe8 },
+	{ CCI_REG8(0x5312), 0x00 },
+	{ CCI_REG8(0x5331), 0x02 },
+	{ CCI_REG8(0x5332), 0x42 },
+	{ CCI_REG8(0x5333), 0x24 },
+	{ CCI_REG8(0x5353), 0x09 },
+	{ CCI_REG8(0x5354), 0x00 },
+	{ CCI_REG8(0x53c1), 0x00 },
+	{ CCI_REG8(0x5414), 0x01 },
+	{ CCI_REG8(0x5480), 0xc1 },
+	{ CCI_REG8(0x548a), 0x04 },
+	{ CCI_REG8(0x54b0), 0x10 },
+	{ CCI_REG8(0x54d1), 0x68 },
+	{ CCI_REG8(0x54d2), 0xff },
+	{ CCI_REG8(0x54d3), 0x1c },
+	{ CCI_REG8(0x54ee), 0x01 },
+	{ CCI_REG8(0x54ef), 0x02 },
+	{ CCI_REG8(0x54f0), 0x04 },
+	{ CCI_REG8(0x5510), 0x03 },
+	{ CCI_REG8(0x5511), 0xe8 },
+	{ CCI_REG8(0x5550), 0x06 },
+	{ CCI_REG8(0x557a), 0x00 },
+	{ CCI_REG8(0x557b), 0x38 },
+	{ CCI_REG8(0x557c), 0x00 },
+	{ CCI_REG8(0x557d), 0x4b },
+	{ CCI_REG8(0x5598), 0x00 },
+	{ CCI_REG8(0x5599), 0x50 },
+	{ CCI_REG8(0x559a), 0x00 },
+	{ CCI_REG8(0x559b), 0x50 },
+	{ CCI_REG8(0x559c), 0x00 },
+	{ CCI_REG8(0x559d), 0x50 },
+	{ CCI_REG8(0x559e), 0x00 },
+	{ CCI_REG8(0x559f), 0x50 },
+	{ CCI_REG8(0x55a0), 0x00 },
+	{ CCI_REG8(0x55a1), 0x50 },
+	{ CCI_REG8(0x55a2), 0x00 },
+	{ CCI_REG8(0x55a3), 0x50 },
+	{ CCI_REG8(0x55a4), 0x00 },
+	{ CCI_REG8(0x55a5), 0x50 },
+	{ CCI_REG8(0x55a6), 0x00 },
+	{ CCI_REG8(0x55a7), 0x50 },
+	{ CCI_REG8(0x55a8), 0x00 },
+	{ CCI_REG8(0x55a9), 0x50 },
+	{ CCI_REG8(0x55aa), 0x00 },
+	{ CCI_REG8(0x55ab), 0x50 },
+	{ CCI_REG8(0x55ac), 0x00 },
+	{ CCI_REG8(0x55ad), 0x50 },
+	{ CCI_REG8(0x55ae), 0x00 },
+	{ CCI_REG8(0x55af), 0x50 },
+	{ CCI_REG8(0x55d0), 0x20 },
+	{ CCI_REG8(0x55d1), 0x20 },
+	{ CCI_REG8(0x55d2), 0x20 },
+	{ CCI_REG8(0x55d3), 0x20 },
+	{ CCI_REG8(0x55d4), 0x20 },
+	{ CCI_REG8(0x55d5), 0x20 },
+	{ CCI_REG8(0x55d6), 0x20 },
+	{ CCI_REG8(0x55d7), 0x20 },
+	{ CCI_REG8(0x55d8), 0x20 },
+	{ CCI_REG8(0x55d9), 0x20 },
+	{ CCI_REG8(0x55da), 0x20 },
+	{ CCI_REG8(0x55db), 0x20 },
+	{ CCI_REG8(0x55dc), 0x20 },
+	{ CCI_REG8(0x55dd), 0x20 },
+	{ CCI_REG8(0x55de), 0x20 },
+	{ CCI_REG8(0x55df), 0x20 },
+	{ CCI_REG8(0x55f0), 0x01 },
+	{ CCI_REG8(0x55f1), 0x02 },
+	{ CCI_REG8(0x55f2), 0x03 },
+	{ CCI_REG8(0x55f3), 0x04 },
+	{ CCI_REG8(0x55f4), 0x05 },
+	{ CCI_REG8(0x55f5), 0x06 },
+	{ CCI_REG8(0x55f6), 0x07 },
+	{ CCI_REG8(0x55f7), 0x08 },
+	{ CCI_REG8(0x55f8), 0x09 },
+	{ CCI_REG8(0x55f9), 0x0a },
+	{ CCI_REG8(0x55fa), 0x0b },
+	{ CCI_REG8(0x55fb), 0x0c },
+	{ CCI_REG8(0x55fc), 0x0d },
+	{ CCI_REG8(0x55fd), 0x0e },
+	{ CCI_REG8(0x55fe), 0x0f },
+	{ CCI_REG8(0x55ff), 0x10 },
+	{ CCI_REG8(0x5600), 0x11 },
+	{ CCI_REG8(0x5601), 0x12 },
+	{ CCI_REG8(0x5602), 0x13 },
+	{ CCI_REG8(0x5603), 0x14 },
+	{ CCI_REG8(0x5604), 0x15 },
+	{ CCI_REG8(0x5605), 0x16 },
+	{ CCI_REG8(0x5606), 0x17 },
+	{ CCI_REG8(0x5607), 0x18 },
+	{ CCI_REG8(0x5608), 0x19 },
+	{ CCI_REG8(0x5609), 0x1a },
+	{ CCI_REG8(0x560a), 0x1b },
+	{ CCI_REG8(0x560b), 0x1c },
+	{ CCI_REG8(0x560c), 0x1d },
+	{ CCI_REG8(0x560d), 0x1e },
+	{ CCI_REG8(0x560e), 0x1f },
+	{ CCI_REG8(0x560f), 0x20 },
+	{ CCI_REG8(0x5631), 0x02 },
+	{ CCI_REG8(0x5632), 0x42 },
+	{ CCI_REG8(0x5633), 0x24 },
+	{ CCI_REG8(0x5653), 0x09 },
+	{ CCI_REG8(0x5654), 0x00 },
+	{ CCI_REG8(0x56c1), 0x00 },
+	{ CCI_REG8(0x5714), 0x01 },
+	{ CCI_REG8(0x5780), 0xc1 },
+	{ CCI_REG8(0x578a), 0x04 },
+	{ CCI_REG8(0x57b0), 0x10 },
+	{ CCI_REG8(0x57d1), 0x68 },
+	{ CCI_REG8(0x57d2), 0xff },
+	{ CCI_REG8(0x57d3), 0x1c },
+	{ CCI_REG8(0x57ee), 0x01 },
+	{ CCI_REG8(0x57ef), 0x02 },
+	{ CCI_REG8(0x57f0), 0x04 },
+	{ CCI_REG8(0x5810), 0x03 },
+	{ CCI_REG8(0x5811), 0xe8 },
+	{ CCI_REG8(0x5850), 0x06 },
+	{ CCI_REG8(0x587a), 0x00 },
+	{ CCI_REG8(0x587b), 0x38 },
+	{ CCI_REG8(0x587c), 0x00 },
+	{ CCI_REG8(0x587d), 0x4b },
+	{ CCI_REG8(0x5898), 0x00 },
+	{ CCI_REG8(0x5899), 0x50 },
+	{ CCI_REG8(0x589a), 0x00 },
+	{ CCI_REG8(0x589b), 0x50 },
+	{ CCI_REG8(0x589c), 0x00 },
+	{ CCI_REG8(0x589d), 0x50 },
+	{ CCI_REG8(0x589e), 0x00 },
+	{ CCI_REG8(0x589f), 0x50 },
+	{ CCI_REG8(0x58a0), 0x00 },
+	{ CCI_REG8(0x58a1), 0x50 },
+	{ CCI_REG8(0x58a2), 0x00 },
+	{ CCI_REG8(0x58a3), 0x50 },
+	{ CCI_REG8(0x58a4), 0x00 },
+	{ CCI_REG8(0x58a5), 0x50 },
+	{ CCI_REG8(0x58a6), 0x00 },
+	{ CCI_REG8(0x58a7), 0x50 },
+	{ CCI_REG8(0x58a8), 0x00 },
+	{ CCI_REG8(0x58a9), 0x50 },
+	{ CCI_REG8(0x58aa), 0x00 },
+	{ CCI_REG8(0x58ab), 0x50 },
+	{ CCI_REG8(0x58ac), 0x00 },
+	{ CCI_REG8(0x58ad), 0x50 },
+	{ CCI_REG8(0x58ae), 0x00 },
+	{ CCI_REG8(0x58af), 0x50 },
+	{ CCI_REG8(0x58d0), 0x20 },
+	{ CCI_REG8(0x58d1), 0x20 },
+	{ CCI_REG8(0x58d2), 0x20 },
+	{ CCI_REG8(0x58d3), 0x20 },
+	{ CCI_REG8(0x58d4), 0x20 },
+	{ CCI_REG8(0x58d5), 0x20 },
+	{ CCI_REG8(0x58d6), 0x20 },
+	{ CCI_REG8(0x58d7), 0x20 },
+	{ CCI_REG8(0x58d8), 0x20 },
+	{ CCI_REG8(0x58d9), 0x20 },
+	{ CCI_REG8(0x58da), 0x20 },
+	{ CCI_REG8(0x58db), 0x20 },
+	{ CCI_REG8(0x58dc), 0x20 },
+	{ CCI_REG8(0x58dd), 0x20 },
+	{ CCI_REG8(0x58de), 0x20 },
+	{ CCI_REG8(0x58df), 0x20 },
+	{ CCI_REG8(0x58f0), 0x01 },
+	{ CCI_REG8(0x58f1), 0x02 },
+	{ CCI_REG8(0x58f2), 0x03 },
+	{ CCI_REG8(0x58f3), 0x04 },
+	{ CCI_REG8(0x58f4), 0x05 },
+	{ CCI_REG8(0x58f5), 0x06 },
+	{ CCI_REG8(0x58f6), 0x07 },
+	{ CCI_REG8(0x58f7), 0x08 },
+	{ CCI_REG8(0x58f8), 0x09 },
+	{ CCI_REG8(0x58f9), 0x0a },
+	{ CCI_REG8(0x58fa), 0x0b },
+	{ CCI_REG8(0x58fb), 0x0c },
+	{ CCI_REG8(0x58fc), 0x0d },
+	{ CCI_REG8(0x58fd), 0x0e },
+	{ CCI_REG8(0x58fe), 0x0f },
+	{ CCI_REG8(0x58ff), 0x10 },
+	{ CCI_REG8(0x5900), 0x11 },
+	{ CCI_REG8(0x5901), 0x12 },
+	{ CCI_REG8(0x5902), 0x13 },
+	{ CCI_REG8(0x5903), 0x14 },
+	{ CCI_REG8(0x5904), 0x15 },
+	{ CCI_REG8(0x5905), 0x16 },
+	{ CCI_REG8(0x5906), 0x17 },
+	{ CCI_REG8(0x5907), 0x18 },
+	{ CCI_REG8(0x5908), 0x19 },
+	{ CCI_REG8(0x5909), 0x1a },
+	{ CCI_REG8(0x590a), 0x1b },
+	{ CCI_REG8(0x590b), 0x1c },
+	{ CCI_REG8(0x590c), 0x1d },
+	{ CCI_REG8(0x590d), 0x1e },
+	{ CCI_REG8(0x590e), 0x1f },
+	{ CCI_REG8(0x590f), 0x20 },
+	{ CCI_REG8(0x5931), 0x02 },
+	{ CCI_REG8(0x5932), 0x42 },
+	{ CCI_REG8(0x5933), 0x24 },
+	{ CCI_REG8(0x5953), 0x09 },
+	{ CCI_REG8(0x5954), 0x00 },
+	{ CCI_REG8(0x5980), 0x3a },
+	{ CCI_REG8(0x5989), 0x84 },
+	{ CCI_REG8(0x59c3), 0x04 },
+	{ CCI_REG8(0x59c4), 0x24 },
+	{ CCI_REG8(0x59c5), 0x40 },
+	{ CCI_REG8(0x59c6), 0x1b },
+	{ CCI_REG8(0x59c7), 0x40 },
+	{ CCI_REG8(0x5a02), 0x0f },
+	{ CCI_REG8(0x5b80), 0x9c },
+	{ CCI_REG8(0x5b81), 0x8c },
+	{ CCI_REG8(0x5b82), 0x77 },
+	{ CCI_REG8(0x5b83), 0x82 },
+	{ CCI_REG8(0x5b84), 0x8b },
+	{ CCI_REG8(0x5b85), 0x88 },
+	{ CCI_REG8(0x5b86), 0x9f },
+	{ CCI_REG8(0x5b87), 0x86 },
+	{ CCI_REG8(0x5b88), 0x88 },
+	{ CCI_REG8(0x5b89), 0x8c },
+	{ CCI_REG8(0x5b8a), 0x80 },
+	{ CCI_REG8(0x5b8b), 0x88 },
+	{ CCI_REG8(0x5b8c), 0xac },
+	{ CCI_REG8(0x5b8d), 0x96 },
+	{ CCI_REG8(0x5b8e), 0x92 },
+	{ CCI_REG8(0x5b8f), 0x84 },
+	{ CCI_REG8(0x5b90), 0x8a },
+	{ CCI_REG8(0x5b91), 0x6c },
+	{ CCI_REG8(0x5b92), 0xa3 },
+	{ CCI_REG8(0x5b93), 0x98 },
+	{ CCI_REG8(0x5b94), 0x75 },
+	{ CCI_REG8(0x5b95), 0x7d },
+	{ CCI_REG8(0x5b96), 0x7c },
+	{ CCI_REG8(0x5b97), 0x6e },
+	{ CCI_REG8(0x5b98), 0x90 },
+	{ CCI_REG8(0x5b99), 0x9b },
+	{ CCI_REG8(0x5b9a), 0x8b },
+	{ CCI_REG8(0x5b9b), 0x76 },
+	{ CCI_REG8(0x5b9c), 0x64 },
+	{ CCI_REG8(0x5b9d), 0x63 },
+	{ CCI_REG8(0x5b9e), 0x99 },
+	{ CCI_REG8(0x5b9f), 0x85 },
+	{ CCI_REG8(0x5ba0), 0x7d },
+	{ CCI_REG8(0x5ba1), 0x7f },
+	{ CCI_REG8(0x5ba2), 0x7b },
+	{ CCI_REG8(0x5ba3), 0x61 },
+	{ CCI_REG8(0x5ba4), 0x6b },
+	{ CCI_REG8(0x5ba5), 0x60 },
+	{ CCI_REG8(0x5ba6), 0x66 },
+	{ CCI_REG8(0x5ba7), 0x69 },
+	{ CCI_REG8(0x5ba8), 0x7e },
+	{ CCI_REG8(0x5ba9), 0x73 },
+	{ CCI_REG8(0x5baa), 0x63 },
+	{ CCI_REG8(0x5bab), 0x61 },
+	{ CCI_REG8(0x5bac), 0x64 },
+	{ CCI_REG8(0x5bad), 0x64 },
+	{ CCI_REG8(0x5bae), 0x79 },
+	{ CCI_REG8(0x5baf), 0x8a },
+	{ CCI_REG8(0x5bb0), 0x54 },
+	{ CCI_REG8(0x5bb1), 0x6e },
+	{ CCI_REG8(0x5bb2), 0x68 },
+	{ CCI_REG8(0x5bb3), 0x7b },
+	{ CCI_REG8(0x5bb4), 0x82 },
+	{ CCI_REG8(0x5bb5), 0xad },
+	{ CCI_REG8(0x5bb6), 0x57 },
+	{ CCI_REG8(0x5bb7), 0x78 },
+	{ CCI_REG8(0x5bb8), 0x74 },
+	{ CCI_REG8(0x5bb9), 0x81 },
+	{ CCI_REG8(0x5bba), 0x9c },
+	{ CCI_REG8(0x5bbb), 0xa7 },
+	{ CCI_REG8(0x5bbc), 0x67 },
+	{ CCI_REG8(0x5bbd), 0x7b },
+	{ CCI_REG8(0x5bbe), 0x70 },
+	{ CCI_REG8(0x5bbf), 0x89 },
+	{ CCI_REG8(0x5bc0), 0x96 },
+	{ CCI_REG8(0x5bc1), 0xac },
+	{ CCI_REG8(0x5bc2), 0x7b },
+	{ CCI_REG8(0x5bc3), 0x65 },
+	{ CCI_REG8(0x5bc4), 0x8a },
+	{ CCI_REG8(0x5bc5), 0x8c },
+	{ CCI_REG8(0x5bc6), 0x9e },
+	{ CCI_REG8(0x5bc7), 0x95 },
+	{ CCI_REG8(0x5bc8), 0x9c },
+	{ CCI_REG8(0x5bc9), 0x9c },
+	{ CCI_REG8(0x5bca), 0x95 },
+	{ CCI_REG8(0x5bcb), 0xab },
+	{ CCI_REG8(0x5bcc), 0x84 },
+	{ CCI_REG8(0x5bcd), 0x89 },
+	{ CCI_REG8(0x5bce), 0x80 },
+	{ CCI_REG8(0x5bcf), 0x9a },
+	{ CCI_REG8(0x5bd0), 0x9c },
+	{ CCI_REG8(0x5bd1), 0x98 },
+	{ CCI_REG8(0x5bd2), 0x8d },
+	{ CCI_REG8(0x5bd3), 0x73 },
+	{ CCI_REG8(0x5bd4), 0x83 },
+	{ CCI_REG8(0x5bd5), 0x86 },
+	{ CCI_REG8(0x5bd6), 0x9e },
+	{ CCI_REG8(0x5bd7), 0x8c },
+	{ CCI_REG8(0x5bd8), 0x78 },
+	{ CCI_REG8(0x5bd9), 0x56 },
+	{ CCI_REG8(0x5bda), 0x8c },
+	{ CCI_REG8(0x5bdb), 0x74 },
+	{ CCI_REG8(0x5bdc), 0x8a },
+	{ CCI_REG8(0x5bdd), 0x79 },
+	{ CCI_REG8(0x5bde), 0x6e },
+	{ CCI_REG8(0x5bdf), 0x5e },
+	{ CCI_REG8(0x5be0), 0x75 },
+	{ CCI_REG8(0x5be1), 0x8b },
+	{ CCI_REG8(0x5be2), 0x8c },
+	{ CCI_REG8(0x5be3), 0x71 },
+	{ CCI_REG8(0x5be4), 0x6f },
+	{ CCI_REG8(0x5be5), 0x51 },
+	{ CCI_REG8(0x5be6), 0x8b },
+	{ CCI_REG8(0x5be7), 0x8d },
+	{ CCI_REG8(0x5be8), 0x8c },
+	{ CCI_REG8(0x5be9), 0x76 },
+	{ CCI_REG8(0x5bea), 0x67 },
+	{ CCI_REG8(0x5beb), 0x69 },
+	{ CCI_REG8(0x5bec), 0x8d },
+	{ CCI_REG8(0x5bed), 0x88 },
+	{ CCI_REG8(0x5bee), 0x8e },
+	{ CCI_REG8(0x5bef), 0x70 },
+	{ CCI_REG8(0x5bf0), 0x84 },
+	{ CCI_REG8(0x5bf1), 0x91 },
+	{ CCI_REG8(0x5bf2), 0x8d },
+	{ CCI_REG8(0x5bf3), 0x9b },
+	{ CCI_REG8(0x5bf4), 0x82 },
+	{ CCI_REG8(0x5bf5), 0x8e },
+	{ CCI_REG8(0x5bf6), 0x9e },
+	{ CCI_REG8(0x5bf7), 0x92 },
+	{ CCI_REG8(0x5bf8), 0x78 },
+	{ CCI_REG8(0x5bf9), 0x8d },
+	{ CCI_REG8(0x5bfa), 0x87 },
+	{ CCI_REG8(0x5bfb), 0x9e },
+	{ CCI_REG8(0x5bfc), 0xab },
+	{ CCI_REG8(0x5bfd), 0xa7 },
+	{ CCI_REG8(0x5bfe), 0x60 },
+	{ CCI_REG8(0x5bff), 0x76 },
+	{ CCI_REG8(0x5c00), 0x7f },
+	{ CCI_REG8(0x5c01), 0x70 },
+	{ CCI_REG8(0x5c02), 0x9e },
+	{ CCI_REG8(0x5c03), 0xa4 },
+	{ CCI_REG8(0x5c04), 0x79 },
+	{ CCI_REG8(0x5c05), 0x72 },
+	{ CCI_REG8(0x5c06), 0x71 },
+	{ CCI_REG8(0x5c07), 0x71 },
+	{ CCI_REG8(0x5c08), 0x87 },
+	{ CCI_REG8(0x5c09), 0x9d },
+	{ CCI_REG8(0x5c0a), 0x7e },
+	{ CCI_REG8(0x5c0b), 0x71 },
+	{ CCI_REG8(0x5c0c), 0x79 },
+	{ CCI_REG8(0x5c0d), 0x7a },
+	{ CCI_REG8(0x5c0e), 0x81 },
+	{ CCI_REG8(0x5c0f), 0x84 },
+	{ CCI_REG8(0x5c10), 0x7a },
+	{ CCI_REG8(0x5c11), 0x67 },
+	{ CCI_REG8(0x5c12), 0x69 },
+	{ CCI_REG8(0x5c13), 0x61 },
+	{ CCI_REG8(0x5c14), 0x54 },
+	{ CCI_REG8(0x5c15), 0x50 },
+	{ CCI_REG8(0x5c16), 0x72 },
+	{ CCI_REG8(0x5c17), 0x6d },
+	{ CCI_REG8(0x5c18), 0x6d },
+	{ CCI_REG8(0x5c19), 0x62 },
+	{ CCI_REG8(0x5c1a), 0x6a },
+	{ CCI_REG8(0x5c1b), 0x69 },
+	{ CCI_REG8(0x5c1c), 0x9c },
+	{ CCI_REG8(0x5c1d), 0x77 },
+	{ CCI_REG8(0x5c1e), 0x67 },
+	{ CCI_REG8(0x5c1f), 0x6b },
+	{ CCI_REG8(0x5c20), 0x57 },
+	{ CCI_REG8(0x5c21), 0x58 },
+	{ CCI_REG8(0x5c22), 0xae },
+	{ CCI_REG8(0x5c23), 0x86 },
+	{ CCI_REG8(0x5c24), 0x83 },
+	{ CCI_REG8(0x5c25), 0x75 },
+	{ CCI_REG8(0x5c26), 0x64 },
+	{ CCI_REG8(0x5c27), 0x5f },
+	{ CCI_REG8(0x5c28), 0x93 },
+	{ CCI_REG8(0x5c29), 0x84 },
+	{ CCI_REG8(0x5c2a), 0x8a },
+	{ CCI_REG8(0x5c2b), 0x71 },
+	{ CCI_REG8(0x5c2c), 0x7b },
+	{ CCI_REG8(0x5c2d), 0x67 },
+	{ CCI_REG8(0x5c2e), 0x99 },
+	{ CCI_REG8(0x5c2f), 0x8d },
+	{ CCI_REG8(0x5c30), 0x8e },
+	{ CCI_REG8(0x5c31), 0x89 },
+	{ CCI_REG8(0x5c32), 0x65 },
+	{ CCI_REG8(0x5c33), 0x79 },
+	{ CCI_REG8(0x5c34), 0x87 },
+	{ CCI_REG8(0x5c35), 0x9d },
+	{ CCI_REG8(0x5c36), 0xaf },
+	{ CCI_REG8(0x5c37), 0xa9 },
+	{ CCI_REG8(0x5c38), 0xaa },
+	{ CCI_REG8(0x5c39), 0x96 },
+	{ CCI_REG8(0x5c3a), 0x89 },
+	{ CCI_REG8(0x5c3b), 0x99 },
+	{ CCI_REG8(0x5c3c), 0x93 },
+	{ CCI_REG8(0x5c3d), 0x97 },
+	{ CCI_REG8(0x5c3e), 0x93 },
+	{ CCI_REG8(0x5c3f), 0x9a },
+	{ CCI_REG8(0x5c40), 0x60 },
+	{ CCI_REG8(0x5c41), 0x76 },
+	{ CCI_REG8(0x5c42), 0x87 },
+	{ CCI_REG8(0x5c43), 0x92 },
+	{ CCI_REG8(0x5c44), 0x99 },
+	{ CCI_REG8(0x5c45), 0x98 },
+	{ CCI_REG8(0x5c46), 0x56 },
+	{ CCI_REG8(0x5c47), 0x67 },
+	{ CCI_REG8(0x5c48), 0x72 },
+	{ CCI_REG8(0x5c49), 0x89 },
+	{ CCI_REG8(0x5c4a), 0x89 },
+	{ CCI_REG8(0x5c4b), 0x86 },
+	{ CCI_REG8(0x5c4c), 0x6e },
+	{ CCI_REG8(0x5c4d), 0x65 },
+	{ CCI_REG8(0x5c4e), 0x8a },
+	{ CCI_REG8(0x5c4f), 0x82 },
+	{ CCI_REG8(0x5c50), 0x8e },
+	{ CCI_REG8(0x5c51), 0x8b },
+	{ CCI_REG8(0x5c52), 0x61 },
+	{ CCI_REG8(0x5c53), 0x72 },
+	{ CCI_REG8(0x5c54), 0x88 },
+	{ CCI_REG8(0x5c55), 0x82 },
+	{ CCI_REG8(0x5c56), 0x80 },
+	{ CCI_REG8(0x5c57), 0x8e },
+	{ CCI_REG8(0x5c58), 0x9f },
+	{ CCI_REG8(0x5c59), 0x81 },
+	{ CCI_REG8(0x5c5a), 0x7c },
+	{ CCI_REG8(0x5c5b), 0x77 },
+	{ CCI_REG8(0x5c5c), 0x71 },
+	{ CCI_REG8(0x5c5d), 0x7f },
+	{ CCI_REG8(0x5c5e), 0x95 },
+	{ CCI_REG8(0x5c5f), 0x9a },
+	{ CCI_REG8(0x5c60), 0x74 },
+	{ CCI_REG8(0x5c61), 0x75 },
+	{ CCI_REG8(0x5c62), 0x79 },
+	{ CCI_REG8(0x5c63), 0x78 },
+	{ CCI_REG8(0x5c64), 0xa6 },
+	{ CCI_REG8(0x5c65), 0x9b },
+	{ CCI_REG8(0x5c66), 0x76 },
+	{ CCI_REG8(0x5c67), 0x7d },
+	{ CCI_REG8(0x5c68), 0x72 },
+	{ CCI_REG8(0x5c69), 0x6c },
+	{ CCI_REG8(0x5c6a), 0xac },
+	{ CCI_REG8(0x5c6b), 0x96 },
+	{ CCI_REG8(0x5c6c), 0x9d },
+	{ CCI_REG8(0x5c6d), 0x86 },
+	{ CCI_REG8(0x5c6e), 0x8b },
+	{ CCI_REG8(0x5c6f), 0x6e },
+	{ CCI_REG8(0x5c70), 0x93 },
+	{ CCI_REG8(0x5c71), 0x81 },
+	{ CCI_REG8(0x5c72), 0x8c },
+	{ CCI_REG8(0x5c73), 0x84 },
+	{ CCI_REG8(0x5c74), 0x9a },
+	{ CCI_REG8(0x5c75), 0x83 },
+	{ CCI_REG8(0x5c76), 0x9c },
+	{ CCI_REG8(0x5c77), 0x8c },
+	{ CCI_REG8(0x5c78), 0x75 },
+	{ CCI_REG8(0x5c79), 0x98 },
+	{ CCI_REG8(0x5c7a), 0x82 },
+	{ CCI_REG8(0x5c7b), 0x9a },
+	{ CCI_REG8(0x5c7c), 0x7a },
+	{ CCI_REG8(0x5c7d), 0x7f },
+	{ CCI_REG8(0x5c7e), 0x8e },
+	{ CCI_REG8(0x5c7f), 0x8b },
+	{ CCI_REG8(0x5c80), 0x81 },
+	{ CCI_REG8(0x5c81), 0x9d },
+	{ CCI_REG8(0x5c82), 0x62 },
+	{ CCI_REG8(0x5c83), 0x79 },
+	{ CCI_REG8(0x5c84), 0x77 },
+	{ CCI_REG8(0x5c85), 0x8b },
+	{ CCI_REG8(0x5c86), 0x92 },
+	{ CCI_REG8(0x5c87), 0x96 },
+	{ CCI_REG8(0x5c88), 0x53 },
+	{ CCI_REG8(0x5c89), 0x7e },
+	{ CCI_REG8(0x5c8a), 0x88 },
+	{ CCI_REG8(0x5c8b), 0x86 },
+	{ CCI_REG8(0x5c8c), 0x9e },
+	{ CCI_REG8(0x5c8d), 0xa6 },
+	{ CCI_REG8(0x5c8e), 0x57 },
+	{ CCI_REG8(0x5c8f), 0x6e },
+	{ CCI_REG8(0x5c90), 0x69 },
+	{ CCI_REG8(0x5c91), 0x78 },
+	{ CCI_REG8(0x5c92), 0x82 },
+	{ CCI_REG8(0x5c93), 0xa0 },
+	{ CCI_REG8(0x5c94), 0x6f },
+	{ CCI_REG8(0x5c95), 0x64 },
+	{ CCI_REG8(0x5c96), 0x61 },
+	{ CCI_REG8(0x5c97), 0x6d },
+	{ CCI_REG8(0x5c98), 0x65 },
+	{ CCI_REG8(0x5c99), 0x71 },
+	{ CCI_REG8(0x5c9a), 0x68 },
+	{ CCI_REG8(0x5c9b), 0x66 },
+	{ CCI_REG8(0x5c9c), 0x61 },
+	{ CCI_REG8(0x5c9d), 0x51 },
+	{ CCI_REG8(0x5c9e), 0x65 },
+	{ CCI_REG8(0x5c9f), 0x67 },
+	{ CCI_REG8(0x5ca0), 0x8a },
+	{ CCI_REG8(0x5ca1), 0x88 },
+	{ CCI_REG8(0x5ca2), 0x89 },
+	{ CCI_REG8(0x5ca3), 0x75 },
+	{ CCI_REG8(0x5ca4), 0x7b },
+	{ CCI_REG8(0x5ca5), 0x6c },
+	{ CCI_REG8(0x5ca6), 0x8a },
+	{ CCI_REG8(0x5ca7), 0x74 },
+	{ CCI_REG8(0x5ca8), 0x88 },
+	{ CCI_REG8(0x5ca9), 0x71 },
+	{ CCI_REG8(0x5caa), 0x6c },
+	{ CCI_REG8(0x5cab), 0x6a },
+	{ CCI_REG8(0x5cac), 0x82 },
+	{ CCI_REG8(0x5cad), 0x77 },
+	{ CCI_REG8(0x5cae), 0x74 },
+	{ CCI_REG8(0x5caf), 0x78 },
+	{ CCI_REG8(0x5cb0), 0x6f },
+	{ CCI_REG8(0x5cb1), 0x5f },
+	{ CCI_REG8(0x5cb2), 0x80 },
+	{ CCI_REG8(0x5cb3), 0x86 },
+	{ CCI_REG8(0x5cb4), 0x9e },
+	{ CCI_REG8(0x5cb5), 0x8f },
+	{ CCI_REG8(0x5cb6), 0x7b },
+	{ CCI_REG8(0x5cb7), 0x50 },
+	{ CCI_REG8(0x5cb8), 0x86 },
+	{ CCI_REG8(0x5cb9), 0x87 },
+	{ CCI_REG8(0x5cba), 0x9d },
+	{ CCI_REG8(0x5cbb), 0x99 },
+	{ CCI_REG8(0x5cbc), 0x82 },
+	{ CCI_REG8(0x5cbd), 0x71 },
+	{ CCI_REG8(0x5cbe), 0x9e },
+	{ CCI_REG8(0x5cbf), 0x99 },
+	{ CCI_REG8(0x5cc0), 0x97 },
+	{ CCI_REG8(0x5cc1), 0xa9 },
+	{ CCI_REG8(0x5cc2), 0x84 },
+	{ CCI_REG8(0x5cc3), 0x82 },
+	{ CCI_REG8(0x5cc4), 0x72 },
+	{ CCI_REG8(0x5cc5), 0x8a },
+	{ CCI_REG8(0x5cc6), 0x72 },
+	{ CCI_REG8(0x5cc7), 0x65 },
+	{ CCI_REG8(0x5cc8), 0x8f },
+	{ CCI_REG8(0x5cc9), 0x84 },
+	{ CCI_REG8(0x5cca), 0x79 },
+	{ CCI_REG8(0x5ccb), 0x73 },
+	{ CCI_REG8(0x5ccc), 0x76 },
+	{ CCI_REG8(0x5ccd), 0x72 },
+	{ CCI_REG8(0x5cce), 0x87 },
+	{ CCI_REG8(0x5ccf), 0x9e },
+	{ CCI_REG8(0x5cd0), 0x6e },
+	{ CCI_REG8(0x5cd1), 0x73 },
+	{ CCI_REG8(0x5cd2), 0x78 },
+	{ CCI_REG8(0x5cd3), 0x7f },
+	{ CCI_REG8(0x5cd4), 0x9a },
+	{ CCI_REG8(0x5cd5), 0xa1 },
+	{ CCI_REG8(0x5cd6), 0x64 },
+	{ CCI_REG8(0x5cd7), 0x89 },
+	{ CCI_REG8(0x5cd8), 0x8d },
+	{ CCI_REG8(0x5cd9), 0x99 },
+	{ CCI_REG8(0x5cda), 0x94 },
+	{ CCI_REG8(0x5cdb), 0xa4 },
+	{ CCI_REG8(0x5cdc), 0x83 },
+	{ CCI_REG8(0x5cdd), 0x85 },
+	{ CCI_REG8(0x5cde), 0x82 },
+	{ CCI_REG8(0x5cdf), 0x8e },
+	{ CCI_REG8(0x5ce0), 0x9e },
+	{ CCI_REG8(0x5ce1), 0x93 },
+	{ CCI_REG8(0x5ce2), 0x87 },
+	{ CCI_REG8(0x5ce3), 0x8f },
+	{ CCI_REG8(0x5ce4), 0x8f },
+	{ CCI_REG8(0x5ce5), 0x73 },
+	{ CCI_REG8(0x5ce6), 0x99 },
+	{ CCI_REG8(0x5ce7), 0x97 },
+	{ CCI_REG8(0x5ce8), 0x85 },
+	{ CCI_REG8(0x5ce9), 0x8b },
+	{ CCI_REG8(0x5cea), 0x8a },
+	{ CCI_REG8(0x5ceb), 0x8f },
+	{ CCI_REG8(0x5cec), 0x7c },
+	{ CCI_REG8(0x5ced), 0x79 },
+	{ CCI_REG8(0x5cee), 0x92 },
+	{ CCI_REG8(0x5cef), 0x84 },
+	{ CCI_REG8(0x5cf0), 0x75 },
+	{ CCI_REG8(0x5cf1), 0x75 },
+	{ CCI_REG8(0x5cf2), 0x78 },
+	{ CCI_REG8(0x5cf3), 0x7b },
+	{ CCI_REG8(0x5cf4), 0xa3 },
+	{ CCI_REG8(0x5cf5), 0x85 },
+	{ CCI_REG8(0x5cf6), 0x86 },
+	{ CCI_REG8(0x5cf7), 0x8e },
+	{ CCI_REG8(0x5cf8), 0x78 },
+	{ CCI_REG8(0x5cf9), 0x52 },
+	{ CCI_REG8(0x5cfa), 0x93 },
+	{ CCI_REG8(0x5cfb), 0x8b },
+	{ CCI_REG8(0x5cfc), 0x7e },
+	{ CCI_REG8(0x5cfd), 0x69 },
+	{ CCI_REG8(0x5cfe), 0x55 },
+	{ CCI_REG8(0x5cff), 0x58 },
+	{ CCI_REG8(0x5d00), 0x7c },
+	{ CCI_REG8(0x5d01), 0x66 },
+	{ CCI_REG8(0x5d02), 0x63 },
+	{ CCI_REG8(0x5d03), 0x63 },
+	{ CCI_REG8(0x5d04), 0x6b },
+	{ CCI_REG8(0x5d05), 0x69 },
+	{ CCI_REG8(0x5d06), 0x66 },
+	{ CCI_REG8(0x5d07), 0x64 },
+	{ CCI_REG8(0x5d08), 0x6f },
+	{ CCI_REG8(0x5d09), 0x65 },
+	{ CCI_REG8(0x5d0a), 0x51 },
+	{ CCI_REG8(0x5d0b), 0x52 },
+	{ CCI_REG8(0x5d0c), 0x67 },
+	{ CCI_REG8(0x5d0d), 0x71 },
+	{ CCI_REG8(0x5d0e), 0x89 },
+	{ CCI_REG8(0x5d0f), 0x8c },
+	{ CCI_REG8(0x5d10), 0x8f },
+	{ CCI_REG8(0x5d11), 0x88 },
+	{ CCI_REG8(0x5d12), 0x6f },
+	{ CCI_REG8(0x5d13), 0x65 },
+	{ CCI_REG8(0x5d14), 0x8a },
+	{ CCI_REG8(0x5d15), 0x8c },
+	{ CCI_REG8(0x5d16), 0x89 },
+	{ CCI_REG8(0x5d17), 0x74 },
+	{ CCI_REG8(0x5d18), 0x50 },
+	{ CCI_REG8(0x5d19), 0x66 },
+	{ CCI_REG8(0x5d1a), 0x7c },
+	{ CCI_REG8(0x5d1b), 0x8a },
+	{ CCI_REG8(0x5d1c), 0x8a },
+	{ CCI_REG8(0x5d1d), 0x80 },
+	{ CCI_REG8(0x5d1e), 0x62 },
+	{ CCI_REG8(0x5d1f), 0x70 },
+	{ CCI_REG8(0x5d20), 0x83 },
+	{ CCI_REG8(0x5d21), 0x9d },
+	{ CCI_REG8(0x5d22), 0x98 },
+	{ CCI_REG8(0x5d23), 0x98 },
+	{ CCI_REG8(0x5d24), 0x88 },
+	{ CCI_REG8(0x5d25), 0x84 },
+	{ CCI_REG8(0x5d26), 0x9c },
+	{ CCI_REG8(0x5d27), 0x91 },
+	{ CCI_REG8(0x5d28), 0x9d },
+	{ CCI_REG8(0x5d29), 0x85 },
+	{ CCI_REG8(0x5d2a), 0x87 },
+	{ CCI_REG8(0x5d2b), 0x99 },
+	{ CCI_REG8(0x5d2c), 0xab },
+	{ CCI_REG8(0x5d2d), 0x95 },
+	{ CCI_REG8(0x5d2e), 0x95 },
+	{ CCI_REG8(0x5d2f), 0x91 },
+	{ CCI_REG8(0x5d30), 0x75 },
+	{ CCI_REG8(0x5d31), 0x81 },
+	{ CCI_REG8(0x5d32), 0x9d },
+	{ CCI_REG8(0x5d33), 0x9d },
+	{ CCI_REG8(0x5d34), 0x71 },
+	{ CCI_REG8(0x5d35), 0x7e },
+	{ CCI_REG8(0x5d36), 0x7a },
+	{ CCI_REG8(0x5d37), 0x70 },
+	{ CCI_REG8(0x5d38), 0x81 },
+	{ CCI_REG8(0x5d39), 0x8b },
+	{ CCI_REG8(0x5d3a), 0x76 },
+	{ CCI_REG8(0x5d3b), 0x7f },
+	{ CCI_REG8(0x5d3c), 0x8b },
+	{ CCI_REG8(0x5d3d), 0x8a },
+	{ CCI_REG8(0x5d3e), 0x84 },
+	{ CCI_REG8(0x5d3f), 0x7c },
+	{ CCI_REG8(0x5d40), 0x63 },
+	{ CCI_REG8(0x5d41), 0x50 },
+	{ CCI_REG8(0x5d42), 0x8d },
+	{ CCI_REG8(0x5d43), 0x75 },
+	{ CCI_REG8(0x5d44), 0x75 },
+	{ CCI_REG8(0x5d45), 0x67 },
+	{ CCI_REG8(0x5d46), 0x6c },
+	{ CCI_REG8(0x5d47), 0x54 },
+	{ CCI_REG8(0x5d48), 0x75 },
+	{ CCI_REG8(0x5d49), 0x74 },
+	{ CCI_REG8(0x5d4a), 0x77 },
+	{ CCI_REG8(0x5d4b), 0x65 },
+	{ CCI_REG8(0x5d4c), 0x57 },
+	{ CCI_REG8(0x5d4d), 0x69 },
+	{ CCI_REG8(0x5d4e), 0x74 },
+	{ CCI_REG8(0x5d4f), 0x75 },
+	{ CCI_REG8(0x5d50), 0x7c },
+	{ CCI_REG8(0x5d51), 0x7a },
+	{ CCI_REG8(0x5d52), 0x56 },
+	{ CCI_REG8(0x5d53), 0x5d },
+	{ CCI_REG8(0x5d54), 0x7f },
+	{ CCI_REG8(0x5d55), 0x65 },
+	{ CCI_REG8(0x5d56), 0x64 },
+	{ CCI_REG8(0x5d57), 0x79 },
+	{ CCI_REG8(0x5d58), 0x86 },
+	{ CCI_REG8(0x5d59), 0x86 },
+	{ CCI_REG8(0x5d5a), 0x77 },
+	{ CCI_REG8(0x5d5b), 0x73 },
+	{ CCI_REG8(0x5d5c), 0x7f },
+	{ CCI_REG8(0x5d5d), 0x8e },
+	{ CCI_REG8(0x5d5e), 0x8e },
+	{ CCI_REG8(0x5d5f), 0x82 },
+	{ CCI_REG8(0x5d60), 0x62 },
+	{ CCI_REG8(0x5d61), 0x64 },
+	{ CCI_REG8(0x5d62), 0x64 },
+	{ CCI_REG8(0x5d63), 0x80 },
+	{ CCI_REG8(0x5d64), 0x9e },
+	{ CCI_REG8(0x5d65), 0xae },
+	{ CCI_REG8(0x5d66), 0x6f },
+	{ CCI_REG8(0x5d67), 0x7e },
+	{ CCI_REG8(0x5d68), 0x89 },
+	{ CCI_REG8(0x5d69), 0x93 },
+	{ CCI_REG8(0x5d6a), 0x91 },
+	{ CCI_REG8(0x5d6b), 0xa9 },
+	{ CCI_REG8(0x5d6c), 0x7e },
+	{ CCI_REG8(0x5d6d), 0x76 },
+	{ CCI_REG8(0x5d6e), 0x86 },
+	{ CCI_REG8(0x5d6f), 0xaa },
+	{ CCI_REG8(0x5d70), 0xa4 },
+	{ CCI_REG8(0x5d71), 0xa9 },
+	{ CCI_REG8(0x5d72), 0x76 },
+	{ CCI_REG8(0x5d73), 0x88 },
+	{ CCI_REG8(0x5d74), 0x93 },
+	{ CCI_REG8(0x5d75), 0xa9 },
+	{ CCI_REG8(0x5d76), 0xbe },
+	{ CCI_REG8(0x5d77), 0xbe },
+	{ CCI_REG8(0x5d78), 0x85 },
+	{ CCI_REG8(0x5d79), 0x81 },
+	{ CCI_REG8(0x5d7a), 0x8b },
+	{ CCI_REG8(0x5d7b), 0x7d },
+	{ CCI_REG8(0x5d7c), 0x65 },
+	{ CCI_REG8(0x5d7d), 0x7c },
+	{ CCI_REG8(0x5d7e), 0x9a },
+	{ CCI_REG8(0x5d7f), 0x87 },
+	{ CCI_REG8(0x5d80), 0x8f },
+	{ CCI_REG8(0x5d81), 0x7f },
+	{ CCI_REG8(0x5d82), 0x73 },
+	{ CCI_REG8(0x5d83), 0x77 },
+	{ CCI_REG8(0x5d84), 0xab },
+	{ CCI_REG8(0x5d85), 0x92 },
+	{ CCI_REG8(0x5d86), 0x86 },
+	{ CCI_REG8(0x5d87), 0x7d },
+	{ CCI_REG8(0x5d88), 0x7f },
+	{ CCI_REG8(0x5d89), 0x64 },
+	{ CCI_REG8(0x5d8a), 0x94 },
+	{ CCI_REG8(0x5d8b), 0x85 },
+	{ CCI_REG8(0x5d8c), 0x70 },
+	{ CCI_REG8(0x5d8d), 0x63 },
+	{ CCI_REG8(0x5d8e), 0x61 },
+	{ CCI_REG8(0x5d8f), 0x61 },
+	{ CCI_REG8(0x5d90), 0x9a },
+	{ CCI_REG8(0x5d91), 0x8f },
+	{ CCI_REG8(0x5d92), 0x64 },
+	{ CCI_REG8(0x5d93), 0x57 },
+	{ CCI_REG8(0x5d94), 0x51 },
+	{ CCI_REG8(0x5d95), 0x6c },
+	{ CCI_REG8(0x5d96), 0x8f },
+	{ CCI_REG8(0x5d97), 0x76 },
+	{ CCI_REG8(0x5d98), 0x6b },
+	{ CCI_REG8(0x5d99), 0x5f },
+	{ CCI_REG8(0x5d9a), 0x5e },
+	{ CCI_REG8(0x5d9b), 0x52 },
+	{ CCI_REG8(0x5d9c), 0x66 },
+	{ CCI_REG8(0x5d9d), 0x7d },
+	{ CCI_REG8(0x5d9e), 0x9e },
+	{ CCI_REG8(0x5d9f), 0x93 },
+	{ CCI_REG8(0x5da0), 0x85 },
+	{ CCI_REG8(0x5da1), 0x74 },
+	{ CCI_REG8(0x5da2), 0x79 },
+	{ CCI_REG8(0x5da3), 0x70 },
+	{ CCI_REG8(0x5da4), 0x77 },
+	{ CCI_REG8(0x5da5), 0x85 },
+	{ CCI_REG8(0x5da6), 0x8b },
+	{ CCI_REG8(0x5da7), 0x7c },
+	{ CCI_REG8(0x5da8), 0x62 },
+	{ CCI_REG8(0x5da9), 0x67 },
+	{ CCI_REG8(0x5daa), 0x79 },
+	{ CCI_REG8(0x5dab), 0x85 },
+	{ CCI_REG8(0x5dac), 0x8e },
+	{ CCI_REG8(0x5dad), 0x86 },
+	{ CCI_REG8(0x5dae), 0x67 },
+	{ CCI_REG8(0x5daf), 0x66 },
+	{ CCI_REG8(0x5db0), 0x67 },
+	{ CCI_REG8(0x5db1), 0x8a },
+	{ CCI_REG8(0x5db2), 0x8c },
+	{ CCI_REG8(0x5db3), 0x9d },
+	{ CCI_REG8(0x5db4), 0x64 },
+	{ CCI_REG8(0x5db5), 0x62 },
+	{ CCI_REG8(0x5db6), 0x78 },
+	{ CCI_REG8(0x5db7), 0x76 },
+	{ CCI_REG8(0x5db8), 0x88 },
+	{ CCI_REG8(0x5db9), 0x80 },
+	{ CCI_REG8(0x5dba), 0x6d },
+	{ CCI_REG8(0x5dbb), 0x63 },
+	{ CCI_REG8(0x5dbc), 0x7a },
+	{ CCI_REG8(0x5dbd), 0x7c },
+	{ CCI_REG8(0x5dbe), 0x8b },
+	{ CCI_REG8(0x5dbf), 0x8f },
+	{ CCI_REG8(0x5dc0), 0x9d },
+	{ CCI_REG8(0x5dc1), 0x98 },
+	{ CCI_REG8(0x5dc2), 0x7c },
+	{ CCI_REG8(0x5dc3), 0x66 },
+	{ CCI_REG8(0x5dc4), 0x66 },
+	{ CCI_REG8(0x5dc5), 0x7d },
+	{ CCI_REG8(0x5dc6), 0x86 },
+	{ CCI_REG8(0x5dc7), 0x8d },
+	{ CCI_REG8(0x5dc8), 0x83 },
+	{ CCI_REG8(0x5dc9), 0x7b },
+	{ CCI_REG8(0x5dca), 0x78 },
+	{ CCI_REG8(0x5dcb), 0x7d },
+	{ CCI_REG8(0x5dcc), 0x9d },
+	{ CCI_REG8(0x5dcd), 0x98 },
+	{ CCI_REG8(0x5dce), 0x84 },
+	{ CCI_REG8(0x5dcf), 0x67 },
+	{ CCI_REG8(0x5dd0), 0x61 },
+	{ CCI_REG8(0x5dd1), 0x56 },
+	{ CCI_REG8(0x5dd2), 0x9f },
+	{ CCI_REG8(0x5dd3), 0x9d },
+	{ CCI_REG8(0x5dd4), 0x92 },
+	{ CCI_REG8(0x5dd5), 0x88 },
+	{ CCI_REG8(0x5dd6), 0x67 },
+	{ CCI_REG8(0x5dd7), 0x59 },
+	{ CCI_REG8(0x5dd8), 0x9d },
+	{ CCI_REG8(0x5dd9), 0xa9 },
+	{ CCI_REG8(0x5dda), 0x97 },
+	{ CCI_REG8(0x5ddb), 0x84 },
+	{ CCI_REG8(0x5ddc), 0x73 },
+	{ CCI_REG8(0x5ddd), 0x62 },
+	{ CCI_REG8(0x5dde), 0xae },
+	{ CCI_REG8(0x5ddf), 0xac },
+	{ CCI_REG8(0x5de0), 0xa9 },
+	{ CCI_REG8(0x5de1), 0x93 },
+	{ CCI_REG8(0x5de2), 0x75 },
+	{ CCI_REG8(0x5de3), 0x7e },
+	{ CCI_REG8(0x5de4), 0x65 },
+	{ CCI_REG8(0x5de5), 0x61 },
+	{ CCI_REG8(0x5de6), 0x7f },
+	{ CCI_REG8(0x5de7), 0x8a },
+	{ CCI_REG8(0x5de8), 0x86 },
+	{ CCI_REG8(0x5de9), 0x87 },
+	{ CCI_REG8(0x5dea), 0x70 },
+	{ CCI_REG8(0x5deb), 0x7d },
+	{ CCI_REG8(0x5dec), 0x7b },
+	{ CCI_REG8(0x5ded), 0x8f },
+	{ CCI_REG8(0x5dee), 0x9a },
+	{ CCI_REG8(0x5def), 0x9f },
+	{ CCI_REG8(0x5df0), 0x7e },
+	{ CCI_REG8(0x5df1), 0x7f },
+	{ CCI_REG8(0x5df2), 0x7e },
+	{ CCI_REG8(0x5df3), 0x86 },
+	{ CCI_REG8(0x5df4), 0x91 },
+	{ CCI_REG8(0x5df5), 0xa3 },
+	{ CCI_REG8(0x5df6), 0x7a },
+	{ CCI_REG8(0x5df7), 0x66 },
+	{ CCI_REG8(0x5df8), 0x62 },
+	{ CCI_REG8(0x5df9), 0x71 },
+	{ CCI_REG8(0x5dfa), 0x9e },
+	{ CCI_REG8(0x5dfb), 0xa3 },
+	{ CCI_REG8(0x5dfc), 0x61 },
+	{ CCI_REG8(0x5dfd), 0x6a },
+	{ CCI_REG8(0x5dfe), 0x54 },
+	{ CCI_REG8(0x5dff), 0x64 },
+	{ CCI_REG8(0x5e00), 0x8d },
+	{ CCI_REG8(0x5e01), 0x9c },
+	{ CCI_REG8(0x5e02), 0x55 },
+	{ CCI_REG8(0x5e03), 0x51 },
+	{ CCI_REG8(0x5e04), 0x5c },
+	{ CCI_REG8(0x5e05), 0x68 },
+	{ CCI_REG8(0x5e06), 0x74 },
+	{ CCI_REG8(0x5e07), 0x83 },
+	{ CCI_REG8(0x5e08), 0x8a },
+	{ CCI_REG8(0x5e09), 0x70 },
+	{ CCI_REG8(0x5e0a), 0x78 },
+	{ CCI_REG8(0x5e0b), 0x66 },
+	{ CCI_REG8(0x5e0c), 0x6f },
+	{ CCI_REG8(0x5e0d), 0x6f },
+	{ CCI_REG8(0x5e0e), 0x8a },
+	{ CCI_REG8(0x5e0f), 0x73 },
+	{ CCI_REG8(0x5e10), 0x77 },
+	{ CCI_REG8(0x5e11), 0x65 },
+	{ CCI_REG8(0x5e12), 0x6a },
+	{ CCI_REG8(0x5e13), 0x60 },
+	{ CCI_REG8(0x5e14), 0x8f },
+	{ CCI_REG8(0x5e15), 0x73 },
+	{ CCI_REG8(0x5e16), 0x77 },
+	{ CCI_REG8(0x5e17), 0x61 },
+	{ CCI_REG8(0x5e18), 0x6e },
+	{ CCI_REG8(0x5e19), 0x57 },
+	{ CCI_REG8(0x5e1a), 0x77 },
+	{ CCI_REG8(0x5e1b), 0x71 },
+	{ CCI_REG8(0x5e1c), 0x86 },
+	{ CCI_REG8(0x5e1d), 0x7b },
+	{ CCI_REG8(0x5e1e), 0x69 },
+	{ CCI_REG8(0x5e1f), 0x5d },
+	{ CCI_REG8(0x5e20), 0x64 },
+	{ CCI_REG8(0x5e21), 0x7d },
+	{ CCI_REG8(0x5e22), 0x83 },
+	{ CCI_REG8(0x5e23), 0x76 },
+	{ CCI_REG8(0x5e24), 0x73 },
+	{ CCI_REG8(0x5e25), 0x79 },
+	{ CCI_REG8(0x5e26), 0x72 },
+	{ CCI_REG8(0x5e27), 0x8e },
+	{ CCI_REG8(0x5e28), 0x9e },
+	{ CCI_REG8(0x5e29), 0x9b },
+	{ CCI_REG8(0x5e2a), 0x7d },
+	{ CCI_REG8(0x5e2b), 0x7f },
+	{ CCI_REG8(0x5e2c), 0x72 },
+	{ CCI_REG8(0x5e2d), 0x8f },
+	{ CCI_REG8(0x5e2e), 0x91 },
+	{ CCI_REG8(0x5e2f), 0xac },
+	{ CCI_REG8(0x5e30), 0xa0 },
+	{ CCI_REG8(0x5e31), 0xa9 },
+	{ CCI_REG8(0x5e32), 0x7b },
+	{ CCI_REG8(0x5e33), 0x75 },
+	{ CCI_REG8(0x5e34), 0x80 },
+	{ CCI_REG8(0x5e35), 0x94 },
+	{ CCI_REG8(0x5e36), 0xa3 },
+	{ CCI_REG8(0x5e37), 0x93 },
+	{ CCI_REG8(0x5e38), 0x6e },
+	{ CCI_REG8(0x5e39), 0x72 },
+	{ CCI_REG8(0x5e3a), 0x88 },
+	{ CCI_REG8(0x5e3b), 0x93 },
+	{ CCI_REG8(0x5e3c), 0x96 },
+	{ CCI_REG8(0x5e3d), 0xa8 },
+	{ CCI_REG8(0x5e3e), 0x60 },
+	{ CCI_REG8(0x5e3f), 0x78 },
+	{ CCI_REG8(0x5e40), 0x64 },
+	{ CCI_REG8(0x5e41), 0x84 },
+	{ CCI_REG8(0x5e42), 0x92 },
+	{ CCI_REG8(0x5e43), 0xaf },
+	{ CCI_REG8(0x5e44), 0x75 },
+	{ CCI_REG8(0x5e45), 0x70 },
+	{ CCI_REG8(0x5e46), 0x7e },
+	{ CCI_REG8(0x5e47), 0x8d },
+	{ CCI_REG8(0x5e48), 0x8e },
+	{ CCI_REG8(0x5e49), 0x8d },
+	{ CCI_REG8(0x5e4a), 0x73 },
+	{ CCI_REG8(0x5e4b), 0x7e },
+	{ CCI_REG8(0x5e4c), 0x65 },
+	{ CCI_REG8(0x5e4d), 0x7c },
+	{ CCI_REG8(0x5e4e), 0x86 },
+	{ CCI_REG8(0x5e4f), 0x83 },
+	{ CCI_REG8(0x5e50), 0x80 },
+	{ CCI_REG8(0x5e51), 0x76 },
+	{ CCI_REG8(0x5e52), 0x69 },
+	{ CCI_REG8(0x5e53), 0x5c },
+	{ CCI_REG8(0x5e54), 0x50 },
+	{ CCI_REG8(0x5e55), 0x69 },
+	{ CCI_REG8(0x5e56), 0x99 },
+	{ CCI_REG8(0x5e57), 0x8f },
+	{ CCI_REG8(0x5e58), 0x7b },
+	{ CCI_REG8(0x5e59), 0x6a },
+	{ CCI_REG8(0x5e5a), 0x55 },
+	{ CCI_REG8(0x5e5b), 0x66 },
+	{ CCI_REG8(0x5e5c), 0xab },
+	{ CCI_REG8(0x5e5d), 0x85 },
+	{ CCI_REG8(0x5e5e), 0x76 },
+	{ CCI_REG8(0x5e5f), 0x60 },
+	{ CCI_REG8(0x5e60), 0x67 },
+	{ CCI_REG8(0x5e61), 0x67 },
+	{ CCI_REG8(0x5e62), 0xab },
+	{ CCI_REG8(0x5e63), 0x92 },
+	{ CCI_REG8(0x5e64), 0x87 },
+	{ CCI_REG8(0x5e65), 0x7c },
+	{ CCI_REG8(0x5e66), 0x7f },
+	{ CCI_REG8(0x5e67), 0x7a },
+	{ CCI_REG8(0x5e68), 0x9a },
+	{ CCI_REG8(0x5e69), 0x84 },
+	{ CCI_REG8(0x5e6a), 0x8d },
+	{ CCI_REG8(0x5e6b), 0x7c },
+	{ CCI_REG8(0x5e6c), 0x76 },
+	{ CCI_REG8(0x5e6d), 0x75 },
+	{ CCI_REG8(0x5e6e), 0x85 },
+	{ CCI_REG8(0x5e6f), 0x81 },
+	{ CCI_REG8(0x5e70), 0x89 },
+	{ CCI_REG8(0x5e71), 0x73 },
+	{ CCI_REG8(0x5e72), 0x78 },
+	{ CCI_REG8(0x5e73), 0x70 },
+	{ CCI_REG8(0x5e74), 0x69 },
+	{ CCI_REG8(0x5e75), 0x63 },
+	{ CCI_REG8(0x5e76), 0x66 },
+	{ CCI_REG8(0x5e77), 0x64 },
+	{ CCI_REG8(0x5e78), 0x71 },
+	{ CCI_REG8(0x5e79), 0x88 },
+	{ CCI_REG8(0x5e7a), 0x62 },
+	{ CCI_REG8(0x5e7b), 0x69 },
+	{ CCI_REG8(0x5e7c), 0x64 },
+	{ CCI_REG8(0x5e7d), 0x7d },
+	{ CCI_REG8(0x5e7e), 0x77 },
+	{ CCI_REG8(0x5e7f), 0x88 },
+	{ CCI_REG8(0x5e80), 0x6b },
+	{ CCI_REG8(0x5e81), 0x6f },
+	{ CCI_REG8(0x5e82), 0x6d },
+	{ CCI_REG8(0x5e83), 0x70 },
+	{ CCI_REG8(0x5e84), 0x74 },
+	{ CCI_REG8(0x5e85), 0x9a },
+	{ CCI_REG8(0x5e86), 0x55 },
+	{ CCI_REG8(0x5e87), 0x6e },
+	{ CCI_REG8(0x5e88), 0x67 },
+	{ CCI_REG8(0x5e89), 0x80 },
+	{ CCI_REG8(0x5e8a), 0x8a },
+	{ CCI_REG8(0x5e8b), 0x8d },
+	{ CCI_REG8(0x5e8c), 0x64 },
+	{ CCI_REG8(0x5e8d), 0x72 },
+	{ CCI_REG8(0x5e8e), 0x7c },
+	{ CCI_REG8(0x5e8f), 0x82 },
+	{ CCI_REG8(0x5e90), 0x8a },
+	{ CCI_REG8(0x5e91), 0x7c },
+	{ CCI_REG8(0x5e92), 0x7b },
+	{ CCI_REG8(0x5e93), 0x7d },
+	{ CCI_REG8(0x5e94), 0x9a },
+	{ CCI_REG8(0x5e95), 0x9b },
+	{ CCI_REG8(0x5e96), 0x86 },
+	{ CCI_REG8(0x5e97), 0x8b },
+	{ CCI_REG8(0x5e98), 0xaf },
+	{ CCI_REG8(0x5e99), 0xac },
+	{ CCI_REG8(0x5e9a), 0xac },
+	{ CCI_REG8(0x5e9b), 0x95 },
+	{ CCI_REG8(0x5e9c), 0x89 },
+	{ CCI_REG8(0x5e9d), 0x7f },
+	{ CCI_REG8(0x5e9e), 0x94 },
+	{ CCI_REG8(0x5e9f), 0xac },
+	{ CCI_REG8(0x5ea0), 0x95 },
+	{ CCI_REG8(0x5ea1), 0x9a },
+	{ CCI_REG8(0x5ea2), 0x71 },
+	{ CCI_REG8(0x5ea3), 0x65 },
+	{ CCI_REG8(0x5ea4), 0x95 },
+	{ CCI_REG8(0x5ea5), 0x91 },
+	{ CCI_REG8(0x5ea6), 0x97 },
+	{ CCI_REG8(0x5ea7), 0x8f },
+	{ CCI_REG8(0x5ea8), 0x78 },
+	{ CCI_REG8(0x5ea9), 0x53 },
+	{ CCI_REG8(0x5eaa), 0x97 },
+	{ CCI_REG8(0x5eab), 0x92 },
+	{ CCI_REG8(0x5eac), 0x9b },
+	{ CCI_REG8(0x5ead), 0x7a },
+	{ CCI_REG8(0x5eae), 0x67 },
+	{ CCI_REG8(0x5eaf), 0x55 },
+	{ CCI_REG8(0x5eb0), 0x86 },
+	{ CCI_REG8(0x5eb1), 0x8c },
+	{ CCI_REG8(0x5eb2), 0x87 },
+	{ CCI_REG8(0x5eb3), 0x7c },
+	{ CCI_REG8(0x5eb4), 0x7b },
+	{ CCI_REG8(0x5eb5), 0x7c },
+	{ CCI_REG8(0x5eb6), 0x87 },
+	{ CCI_REG8(0x5eb7), 0x87 },
+	{ CCI_REG8(0x5eb8), 0x72 },
+	{ CCI_REG8(0x5eb9), 0x79 },
+	{ CCI_REG8(0x5eba), 0x61 },
+	{ CCI_REG8(0x5ebb), 0x7a },
+	{ CCI_REG8(0x5ebc), 0x6b },
+	{ CCI_REG8(0x5ebd), 0x56 },
+	{ CCI_REG8(0x5ebe), 0x5c },
+	{ CCI_REG8(0x5ebf), 0x6b },
+	{ CCI_REG8(0x5ec0), 0x77 },
+	{ CCI_REG8(0x5ec1), 0x86 },
+	{ CCI_REG8(0x5ec2), 0x62 },
+	{ CCI_REG8(0x5ec3), 0x6a },
+	{ CCI_REG8(0x5ec4), 0x55 },
+	{ CCI_REG8(0x5ec5), 0x65 },
+	{ CCI_REG8(0x5ec6), 0x8d },
+	{ CCI_REG8(0x5ec7), 0x9e },
+	{ CCI_REG8(0x5ec8), 0x67 },
+	{ CCI_REG8(0x5ec9), 0x66 },
+	{ CCI_REG8(0x5eca), 0x62 },
+	{ CCI_REG8(0x5ecb), 0x71 },
+	{ CCI_REG8(0x5ecc), 0x99 },
+	{ CCI_REG8(0x5ecd), 0xad },
+	{ CCI_REG8(0x5ece), 0x7e },
+	{ CCI_REG8(0x5ecf), 0x7e },
+	{ CCI_REG8(0x5ed0), 0x7e },
+	{ CCI_REG8(0x5ed1), 0x87 },
+	{ CCI_REG8(0x5ed2), 0x90 },
+	{ CCI_REG8(0x5ed3), 0xa2 },
+	{ CCI_REG8(0x5ed4), 0x76 },
+	{ CCI_REG8(0x5ed5), 0x71 },
+	{ CCI_REG8(0x5ed6), 0x78 },
+	{ CCI_REG8(0x5ed7), 0x8c },
+	{ CCI_REG8(0x5ed8), 0x99 },
+	{ CCI_REG8(0x5ed9), 0x9f },
+	{ CCI_REG8(0x5eda), 0x72 },
+	{ CCI_REG8(0x5edb), 0x7b },
+	{ CCI_REG8(0x5edc), 0x7d },
+	{ CCI_REG8(0x5edd), 0x8b },
+	{ CCI_REG8(0x5ede), 0x85 },
+	{ CCI_REG8(0x5edf), 0x99 },
+	{ CCI_REG8(0x5f00), 0x29 },
+	{ CCI_REG8(0x5f02), 0x04 },
+	{ CCI_REG8(0x5f2d), 0x28 },
+	{ CCI_REG8(0x5f2e), 0x28 },
+};
+
+static const struct cci_reg_sequence ov64b40_4624_3472[] = {
+	{ CCI_REG8(0x0303), 0x03 },
+	{ CCI_REG8(0x0304), 0x01 },
+	{ CCI_REG8(0x0305), 0x62 },
+	{ CCI_REG8(0x0307), 0x00 },
+	{ CCI_REG8(0x0324), 0x01 },
+	{ CCI_REG8(0x0325), 0x80 },
+	{ CCI_REG8(0x0326), 0xc3 },
+	{ CCI_REG8(0x0327), 0x03 },
+	{ CCI_REG8(0x032f), 0xa1 },
+	{ CCI_REG8(0x3501), 0x0f },
+	{ CCI_REG8(0x3502), 0xe0 },
+	{ CCI_REG8(0x3608), 0x8a },
+	{ CCI_REG8(0x360b), 0x50 },
+	{ CCI_REG8(0x360d), 0x05 },
+	{ CCI_REG8(0x3622), 0x0a },
+	{ CCI_REG8(0x3623), 0x08 },
+	{ CCI_REG8(0x3659), 0x48 },
+	{ CCI_REG8(0x3684), 0x8b },
+	{ CCI_REG8(0x3699), 0x19 },
+	{ CCI_REG8(0x369a), 0x00 },
+	{ CCI_REG8(0x3701), 0x40 },
+	{ CCI_REG8(0x3702), 0x4b },
+	{ CCI_REG8(0x3709), 0xf8 },
+	{ CCI_REG8(0x3712), 0x50 },
+	{ CCI_REG8(0x3714), 0x67 },
+	{ CCI_REG8(0x3723), 0x00 },
+	{ CCI_REG8(0x3724), 0x0d },
+	{ CCI_REG8(0x3770), 0x19 },
+	{ CCI_REG8(0x3774), 0x1a },
+	{ CCI_REG8(0x3791), 0x38 },
+	{ CCI_REG8(0x3797), 0x80 },
+	{ CCI_REG8(0x3800), 0x00 },
+	{ CCI_REG8(0x3801), 0x00 },
+	{ CCI_REG8(0x3802), 0x00 },
+	{ CCI_REG8(0x3803), 0x00 },
+	{ CCI_REG8(0x3804), 0x24 },
+	{ CCI_REG8(0x3805), 0x3f },
+	{ CCI_REG8(0x3806), 0x1b },
+	{ CCI_REG8(0x3807), 0x3f },
+	{ CCI_REG8(0x3808), 0x12 },
+	{ CCI_REG8(0x3809), 0x10 },
+	{ CCI_REG8(0x380a), 0x0d },
+	{ CCI_REG8(0x380b), 0x90 },
+	{ CCI_REG8(0x380c), 0x03 },
+	{ CCI_REG8(0x380d), 0xa8 },
+	{ CCI_REG8(0x380e), 0x10 },
+	{ CCI_REG8(0x380f), 0x06 },
+	{ CCI_REG8(0x3811), 0x09 },
+	{ CCI_REG8(0x3813), 0x08 },
+	{ CCI_REG8(0x3815), 0x11 },
+	{ CCI_REG8(0x381a), 0x0f },
+	{ CCI_REG8(0x381b), 0xfe },
+	{ CCI_REG8(0x381c), 0x01 },
+	{ CCI_REG8(0x381d), 0xd4 },
+	{ CCI_REG8(0x3820), 0x42 },
+	{ CCI_REG8(0x3821), 0x04 },
+	{ CCI_REG8(0x3822), 0x10 },
+	{ CCI_REG8(0x3824), 0x01 },
+	{ CCI_REG8(0x3825), 0xd4 },
+	{ CCI_REG8(0x3826), 0x0f },
+	{ CCI_REG8(0x3827), 0xfe },
+	{ CCI_REG8(0x3830), 0x0b },
+	{ CCI_REG8(0x3837), 0x0c },
+	{ CCI_REG8(0x3889), 0x60 },
+	{ CCI_REG8(0x388b), 0x18 },
+	{ CCI_REG8(0x388c), 0x23 },
+	{ CCI_REG8(0x388d), 0x80 },
+	{ CCI_REG8(0x388e), 0x0d },
+	{ CCI_REG8(0x388f), 0x70 },
+	{ CCI_REG8(0x3909), 0xde },
+	{ CCI_REG8(0x390f), 0x54 },
+	{ CCI_REG8(0x3920), 0x88 },
+	{ CCI_REG8(0x3921), 0x00 },
+	{ CCI_REG8(0x3924), 0x84 },
+	{ CCI_REG8(0x3925), 0x0f },
+	{ CCI_REG8(0x3926), 0xfe },
+	{ CCI_REG8(0x3973), 0x02 },
+	{ CCI_REG8(0x3978), 0x02 },
+	{ CCI_REG8(0x399b), 0x4b },
+	{ CCI_REG8(0x39af), 0x01 },
+	{ CCI_REG8(0x39b7), 0x01 },
+	{ CCI_REG8(0x39bf), 0x01 },
+	{ CCI_REG8(0x39cd), 0x18 },
+	{ CCI_REG8(0x39ce), 0x48 },
+	{ CCI_REG8(0x39cf), 0x48 },
+	{ CCI_REG8(0x39d0), 0x48 },
+	{ CCI_REG8(0x39d2), 0x1b },
+	{ CCI_REG8(0x39d3), 0x4b },
+	{ CCI_REG8(0x39d4), 0x4b },
+	{ CCI_REG8(0x39d5), 0x4b },
+	{ CCI_REG8(0x39fd), 0x01 },
+	{ CCI_REG8(0x39f5), 0x04 },
+	{ CCI_REG8(0x39f6), 0x1a },
+	{ CCI_REG8(0x39f7), 0x00 },
+	{ CCI_REG8(0x39f8), 0x00 },
+	{ CCI_REG8(0x39f9), 0x00 },
+	{ CCI_REG8(0x39ff), 0x00 },
+	{ CCI_REG8(0x3a01), 0x03 },
+	{ CCI_REG8(0x3a03), 0x00 },
+	{ CCI_REG8(0x3a05), 0x05 },
+	{ CCI_REG8(0x3a2d), 0x00 },
+	{ CCI_REG8(0x3a2e), 0x00 },
+	{ CCI_REG8(0x3a2f), 0x20 },
+	{ CCI_REG8(0x3a30), 0x20 },
+	{ CCI_REG8(0x3a4b), 0x08 },
+	{ CCI_REG8(0x3a4c), 0x08 },
+	{ CCI_REG8(0x3a4d), 0x08 },
+	{ CCI_REG8(0x3a4e), 0x08 },
+	{ CCI_REG8(0x3a4f), 0x08 },
+	{ CCI_REG8(0x3a50), 0x08 },
+	{ CCI_REG8(0x3a51), 0x08 },
+	{ CCI_REG8(0x3a52), 0x08 },
+	{ CCI_REG8(0x3a57), 0x15 },
+	{ CCI_REG8(0x3a58), 0x00 },
+	{ CCI_REG8(0x3a59), 0x00 },
+	{ CCI_REG8(0x3a5a), 0x01 },
+	{ CCI_REG8(0x3a73), 0x13 },
+	{ CCI_REG8(0x3a74), 0x10 },
+	{ CCI_REG8(0x3a75), 0x0b },
+	{ CCI_REG8(0x3a76), 0x00 },
+	{ CCI_REG8(0x3a77), 0x13 },
+	{ CCI_REG8(0x3a78), 0x10 },
+	{ CCI_REG8(0x3a79), 0x0b },
+	{ CCI_REG8(0x3a7a), 0x00 },
+	{ CCI_REG8(0x3a7d), 0x60 },
+	{ CCI_REG8(0x4016), 0x0f },
+	{ CCI_REG8(0x4018), 0x07 },
+	{ CCI_REG8(0x401f), 0x00 },
+	{ CCI_REG8(0x45c0), 0x61 },
+	{ CCI_REG8(0x45cb), 0x30 },
+	{ CCI_REG8(0x4641), 0x47 },
+	{ CCI_REG8(0x4643), 0x0c },
+	{ CCI_REG8(0x480e), 0x04 },
+	{ CCI_REG8(0x4837), 0x09 },
+	{ CCI_REG8(0x4916), 0x0f },
+	{ CCI_REG8(0x4918), 0x07 },
+	{ CCI_REG8(0x491f), 0x00 },
+	{ CCI_REG8(0x4a16), 0x0f },
+	{ CCI_REG8(0x4a18), 0x07 },
+	{ CCI_REG8(0x4a1f), 0x00 },
+	{ CCI_REG8(0x5000), 0xf7 },
+	{ CCI_REG8(0x5001), 0x01 },
+	{ CCI_REG8(0x5002), 0xf7 },
+	{ CCI_REG8(0x5068), 0x02 },
+	{ CCI_REG8(0x51b0), 0x00 },
+	{ CCI_REG8(0x51d2), 0xff },
+	{ CCI_REG8(0x51d9), 0x10 },
+	{ CCI_REG8(0x51de), 0x02 },
+	{ CCI_REG8(0x51df), 0x06 },
+	{ CCI_REG8(0x51e0), 0x0a },
+	{ CCI_REG8(0x51e1), 0x0e },
+	{ CCI_REG8(0x5202), 0x0d },
+	{ CCI_REG8(0x5203), 0xa0 },
+	{ CCI_REG8(0x5205), 0x10 },
+	{ CCI_REG8(0x5207), 0x10 },
+	{ CCI_REG8(0x5208), 0x12 },
+	{ CCI_REG8(0x5209), 0x00 },
+	{ CCI_REG8(0x520a), 0x0d },
+	{ CCI_REG8(0x520b), 0x80 },
+	{ CCI_REG8(0x520d), 0x08 },
+	{ CCI_REG8(0x5250), 0x06 },
+	{ CCI_REG8(0x5331), 0x02 },
+	{ CCI_REG8(0x5332), 0x42 },
+	{ CCI_REG8(0x5333), 0x24 },
+	{ CCI_REG8(0x54d2), 0xff },
+	{ CCI_REG8(0x5550), 0x06 },
+	{ CCI_REG8(0x57d2), 0xff },
+	{ CCI_REG8(0x5850), 0x06 },
+};
+
+struct ov64b40_reglist {
+	unsigned int num_regs;
+	const struct cci_reg_sequence *regvals;
+};
+
+static struct ov64b40_mode {
+	unsigned int width;
+	unsigned int height;
+	struct ov64b40_timings {
+		unsigned int vts;
+		unsigned int ppl;
+	} timings_default;
+	unsigned long pixel_rate;
+	const struct ov64b40_reglist reglist;
+	struct v4l2_rect analogue_crop;
+	struct v4l2_rect digital_crop;
+} ov64b40_modes[] = {
+	{
+		.width = 4624,
+		.height = 3472,
+		.timings_default = {
+			.vts = 4102,
+			.ppl = 936,
+		},
+		.pixel_rate = 921600000,
+		.reglist = {
+			.num_regs = ARRAY_SIZE(ov64b40_4624_3472),
+			.regvals = ov64b40_4624_3472,
+		},
+		.analogue_crop = {
+			.left = 0,
+			.top = 0,
+			.width = 9280,
+			.height = 6976,
+		},
+		.digital_crop = {
+			.left = 9,
+			.top = 8,
+			.width = 4624,
+			.height = 3472,
+		},
+	},
+};
+
+struct ov64b40 {
+	struct device *dev;
+
+	struct v4l2_subdev sd;
+	struct media_pad pad;
+
+	struct regmap *cci;
+
+	struct ov64b40_mode *mode;
+
+	struct clk *xclk;
+
+	struct gpio_desc *reset_gpio;
+	struct regulator_bulk_data supplies[ARRAY_SIZE(ov64b40_supply_names)];
+
+	struct v4l2_ctrl_handler ctrl_handler;
+	struct v4l2_ctrl *pixel_rate;
+	struct v4l2_ctrl *exposure;
+	struct v4l2_ctrl *vblank;
+	struct v4l2_ctrl *hblank;
+	struct v4l2_ctrl *vflip;
+	struct v4l2_ctrl *hflip;
+};
+
+static inline struct ov64b40 *sd_to_ov64b40(struct v4l2_subdev *sd)
+{
+	return container_of_const(sd, struct ov64b40, sd);
+}
+
+static int ov64b40_program_geometry(struct ov64b40 *ov64b40)
+{
+	struct ov64b40_mode *mode = ov64b40->mode;
+	struct v4l2_rect *anacrop = &mode->analogue_crop;
+	struct v4l2_rect *digicrop = &mode->digital_crop;
+	const struct ov64b40_timings *timings;
+	int ret = 0;
+
+	/* Analogue crop. */
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRL0,
+		  anacrop->left, &ret);
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRL2,
+		  anacrop->top, &ret);
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRL4,
+		  anacrop->width + anacrop->left - 1, &ret);
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRL6,
+		  anacrop->height + anacrop->top - 1, &ret);
+
+	/* ISP windowing. */
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRL10,
+		  digicrop->left, &ret);
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRL12,
+		  digicrop->top, &ret);
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRL8,
+		  digicrop->width, &ret);
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRLA,
+		  digicrop->height, &ret);
+
+	/* Total timings. */
+	timings = &mode->timings_default;
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRLC, timings->ppl, &ret);
+	cci_write(ov64b40->cci, OV64B40_REG_TIMING_CTRLE, timings->vts, &ret);
+
+	return ret;
+}
+
+static int ov64b40_start_streaming(struct ov64b40 *ov64b40,
+				   struct v4l2_subdev_state *state)
+{
+	const struct ov64b40_reglist *reglist = &ov64b40->mode->reglist;
+	const struct ov64b40_timings *timings;
+	unsigned long delay;
+	int ret;
+
+	ret = pm_runtime_resume_and_get(ov64b40->dev);
+	if (ret < 0)
+		return ret;
+
+	ret = cci_multi_reg_write(ov64b40->cci, ov64b40_init,
+				  ARRAY_SIZE(ov64b40_init), NULL);
+	if (ret)
+		goto error_power_off;
+
+	ret = cci_multi_reg_write(ov64b40->cci, reglist->regvals,
+				  reglist->num_regs, NULL);
+	if (ret)
+		goto error_power_off;
+
+	ret = ov64b40_program_geometry(ov64b40);
+	if (ret)
+		goto error_power_off;
+
+	ret =  __v4l2_ctrl_handler_setup(&ov64b40->ctrl_handler);
+	if (ret)
+		goto error_power_off;
+
+	ret = cci_write(ov64b40->cci, OV64B40_REG_SMIA,
+			OV64B40_REG_SMIA_STREAMING, NULL);
+	if (ret)
+		goto error_power_off;
+
+	/* Link frequency and flips cannot change while streaming. */
+	__v4l2_ctrl_grab(ov64b40->vflip, true);
+	__v4l2_ctrl_grab(ov64b40->hflip, true);
+
+	/* delay: max(4096 xclk pulses, 150usec) + exposure time */
+	timings = &ov64b40->mode->timings_default;
+	delay = DIV_ROUND_UP(4096, OV64B40_XCLK_FREQ / 1000 / 1000);
+	delay = max(delay, 150ul);
+
+	/* The sensor has an internal x8 multiplier on the line length. */
+	delay += DIV_ROUND_UP(timings->ppl * 8 * ov64b40->exposure->cur.val,
+			      ov64b40->mode->pixel_rate / 1000 / 1000);
+	fsleep(delay);
+
+	return 0;
+
+error_power_off:
+	pm_runtime_mark_last_busy(ov64b40->dev);
+	pm_runtime_put_autosuspend(ov64b40->dev);
+
+	return ret;
+}
+
+static int ov64b40_stop_streaming(struct ov64b40 *ov64b40,
+				  struct v4l2_subdev_state *state)
+{
+	cci_update_bits(ov64b40->cci, OV64B40_REG_SMIA, BIT(0), 0, NULL);
+	pm_runtime_mark_last_busy(ov64b40->dev);
+	pm_runtime_put_autosuspend(ov64b40->dev);
+
+	__v4l2_ctrl_grab(ov64b40->vflip, false);
+	__v4l2_ctrl_grab(ov64b40->hflip, false);
+
+	return 0;
+}
+
+static int ov64b40_set_stream(struct v4l2_subdev *sd, int enable)
+{
+	struct ov64b40 *ov64b40 = sd_to_ov64b40(sd);
+	struct v4l2_subdev_state *state;
+	int ret;
+
+	state = v4l2_subdev_lock_and_get_active_state(sd);
+	if (enable)
+		ret = ov64b40_start_streaming(ov64b40, state);
+	else
+		ret = ov64b40_stop_streaming(ov64b40, state);
+	v4l2_subdev_unlock_state(state);
+
+	return ret;
+}
+
+static const struct v4l2_subdev_video_ops ov64b40_video_ops = {
+	.s_stream = ov64b40_set_stream,
+};
+
+static u32 ov64b40_mbus_code(struct ov64b40 *ov64b40)
+{
+	unsigned int index = ov64b40->hflip->val << 1 | ov64b40->vflip->val;
+
+	return ov64b40_mbus_codes[index];
+}
+
+static void ov64b40_update_pad_fmt(struct ov64b40 *ov64b40,
+				   struct ov64b40_mode *mode,
+				   struct v4l2_mbus_framefmt *fmt)
+{
+	fmt->code = ov64b40_mbus_code(ov64b40);
+	fmt->width = mode->width;
+	fmt->height = mode->height;
+	fmt->field = V4L2_FIELD_NONE;
+	fmt->colorspace = V4L2_COLORSPACE_RAW;
+	fmt->quantization = V4L2_QUANTIZATION_FULL_RANGE;
+	fmt->xfer_func = V4L2_XFER_FUNC_NONE;
+	fmt->ycbcr_enc = V4L2_YCBCR_ENC_601;
+}
+
+static int ov64b40_init_state(struct v4l2_subdev *sd,
+			      struct v4l2_subdev_state *state)
+{
+	struct ov64b40 *ov64b40 = sd_to_ov64b40(sd);
+	struct v4l2_mbus_framefmt *format;
+	struct v4l2_rect *crop;
+
+	format = v4l2_subdev_state_get_format(state, 0);
+	ov64b40_update_pad_fmt(ov64b40, &ov64b40_modes[0], format);
+
+	crop = v4l2_subdev_state_get_crop(state, 0);
+	crop->top = OV64B40_PIXEL_ARRAY_TOP;
+	crop->left = OV64B40_PIXEL_ARRAY_LEFT;
+	crop->width = OV64B40_PIXEL_ARRAY_WIDTH;
+	crop->height = OV64B40_PIXEL_ARRAY_HEIGHT;
+
+	return 0;
+}
+
+static int ov64b40_enum_mbus_code(struct v4l2_subdev *sd,
+				  struct v4l2_subdev_state *state,
+				  struct v4l2_subdev_mbus_code_enum *code)
+{
+	struct ov64b40 *ov64b40 = sd_to_ov64b40(sd);
+
+	if (code->index)
+		return -EINVAL;
+
+	code->code = ov64b40_mbus_code(ov64b40);
+
+	return 0;
+}
+
+static int ov64b40_enum_frame_size(struct v4l2_subdev *sd,
+				   struct v4l2_subdev_state *state,
+				   struct v4l2_subdev_frame_size_enum *fse)
+{
+	struct ov64b40 *ov64b40 = sd_to_ov64b40(sd);
+	struct ov64b40_mode *mode;
+	u32 code;
+
+	if (fse->index >= ARRAY_SIZE(ov64b40_modes))
+		return -EINVAL;
+
+	code = ov64b40_mbus_code(ov64b40);
+	if (fse->code != code)
+		return -EINVAL;
+
+	mode = &ov64b40_modes[fse->index];
+	fse->min_width = mode->width;
+	fse->max_width = mode->width;
+	fse->min_height = mode->height;
+	fse->max_height = mode->height;
+
+	return 0;
+}
+
+static int ov64b40_get_selection(struct v4l2_subdev *sd,
+				 struct v4l2_subdev_state *state,
+				 struct v4l2_subdev_selection *sel)
+{
+	switch (sel->target) {
+	case V4L2_SEL_TGT_CROP:
+		sel->r = *v4l2_subdev_state_get_crop(state, 0);
+
+		return 0;
+
+	case V4L2_SEL_TGT_NATIVE_SIZE:
+		sel->r.top = 0;
+		sel->r.left = 0;
+		sel->r.width = OV64B40_NATIVE_WIDTH;
+		sel->r.height = OV64B40_NATIVE_HEIGHT;
+
+		return 0;
+
+	case V4L2_SEL_TGT_CROP_DEFAULT:
+	case V4L2_SEL_TGT_CROP_BOUNDS:
+		sel->r.top = OV64B40_PIXEL_ARRAY_TOP;
+		sel->r.left = OV64B40_PIXEL_ARRAY_LEFT;
+		sel->r.width = OV64B40_PIXEL_ARRAY_WIDTH;
+		sel->r.height = OV64B40_PIXEL_ARRAY_HEIGHT;
+
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+static int ov64b40_set_format(struct v4l2_subdev *sd,
+			      struct v4l2_subdev_state *state,
+			      struct v4l2_subdev_format *fmt)
+{
+	struct ov64b40 *ov64b40 = sd_to_ov64b40(sd);
+	struct v4l2_mbus_framefmt *format;
+	struct ov64b40_mode *mode;
+
+	mode = v4l2_find_nearest_size(ov64b40_modes,
+				      ARRAY_SIZE(ov64b40_modes),
+				      width, height,
+				      fmt->format.width, fmt->format.height);
+
+	ov64b40_update_pad_fmt(ov64b40, mode, &fmt->format);
+
+	format = v4l2_subdev_state_get_format(state, 0);
+	if (ov64b40->mode == mode && format->code == fmt->format.code)
+		return 0;
+
+	if (fmt->which == V4L2_SUBDEV_FORMAT_ACTIVE) {
+		const struct ov64b40_timings *timings;
+		int vblank_max, vblank_def;
+		int hblank_val;
+		int exp_max;
+
+		ov64b40->mode = mode;
+		*v4l2_subdev_state_get_crop(state, 0) = mode->analogue_crop;
+
+		/* Update control limits according to the new mode. */
+		__v4l2_ctrl_modify_range(ov64b40->pixel_rate, mode->pixel_rate,
+					 mode->pixel_rate, 1, mode->pixel_rate);
+		timings = &mode->timings_default;
+		vblank_max = OV64B40_VTS_MAX - mode->height;
+		vblank_def = timings->vts - mode->height;
+		__v4l2_ctrl_modify_range(ov64b40->vblank, OV64B40_VBLANK_MIN,
+					 vblank_max, 1, vblank_def);
+		__v4l2_ctrl_s_ctrl(ov64b40->vblank, vblank_def);
+
+		exp_max = timings->vts - OV64B40_EXPOSURE_MARGIN;
+		__v4l2_ctrl_modify_range(ov64b40->exposure,
+					 OV64B40_EXPOSURE_MIN, exp_max,
+					 1, OV64B40_EXPOSURE_MIN);
+
+		hblank_val = timings->ppl * 8 - mode->width;
+		__v4l2_ctrl_modify_range(ov64b40->hblank,
+					 hblank_val, hblank_val, 1, hblank_val);
+	}
+
+	*format = fmt->format;
+
+	return 0;
+}
+
+static const struct v4l2_subdev_pad_ops ov64b40_pad_ops = {
+	.enum_mbus_code = ov64b40_enum_mbus_code,
+	.enum_frame_size = ov64b40_enum_frame_size,
+	.get_fmt = v4l2_subdev_get_fmt,
+	.set_fmt = ov64b40_set_format,
+	.get_selection = ov64b40_get_selection,
+};
+
+static const struct v4l2_subdev_ops ov64b40_subdev_ops = {
+	.video = &ov64b40_video_ops,
+	.pad = &ov64b40_pad_ops,
+};
+
+static const struct v4l2_subdev_internal_ops ov64b40_internal_ops = {
+	.init_state = ov64b40_init_state,
+};
+
+static int ov64b40_power_on(struct device *dev)
+{
+	struct v4l2_subdev *sd = dev_get_drvdata(dev);
+	struct ov64b40 *ov64b40 = sd_to_ov64b40(sd);
+	int ret;
+
+	ret = clk_prepare_enable(ov64b40->xclk);
+	if (ret)
+		return ret;
+
+	ret = regulator_bulk_enable(ARRAY_SIZE(ov64b40_supply_names),
+				    ov64b40->supplies);
+	if (ret) {
+		clk_disable_unprepare(ov64b40->xclk);
+		dev_err(dev, "Failed to enable regulators: %d\n", ret);
+		return ret;
+	}
+
+	gpiod_set_value_cansleep(ov64b40->reset_gpio, 0);
+
+	fsleep(5000);
+
+	return 0;
+}
+
+static int ov64b40_power_off(struct device *dev)
+{
+	struct v4l2_subdev *sd = dev_get_drvdata(dev);
+	struct ov64b40 *ov64b40 = sd_to_ov64b40(sd);
+
+	gpiod_set_value_cansleep(ov64b40->reset_gpio, 1);
+	regulator_bulk_disable(ARRAY_SIZE(ov64b40_supply_names),
+			       ov64b40->supplies);
+	clk_disable_unprepare(ov64b40->xclk);
+
+	return 0;
+}
+
+static int ov64b40_set_ctrl(struct v4l2_ctrl *ctrl)
+{
+	struct ov64b40 *ov64b40 = container_of(ctrl->handler, struct ov64b40,
+					       ctrl_handler);
+	int pm_status;
+	int ret = 0;
+
+	if (ctrl->id == V4L2_CID_VBLANK) {
+		int exp_max = ov64b40->mode->height + ctrl->val
+			    - OV64B40_EXPOSURE_MARGIN;
+		int exp_val = min(ov64b40->exposure->cur.val, exp_max);
+
+		__v4l2_ctrl_modify_range(ov64b40->exposure,
+					 ov64b40->exposure->minimum,
+					 exp_max, 1, exp_val);
+	}
+
+	pm_status = pm_runtime_get_if_active(ov64b40->dev);
+	if (!pm_status)
+		return 0;
+
+	switch (ctrl->id) {
+	case V4L2_CID_EXPOSURE:
+		ret = cci_write(ov64b40->cci, OV64B40_REG_MEC_LONG_EXPO,
+				ctrl->val, NULL);
+		break;
+	case V4L2_CID_ANALOGUE_GAIN:
+		ret = cci_write(ov64b40->cci, OV64B40_REG_MEC_LONG_GAIN,
+				ctrl->val << 1, NULL);
+		break;
+	case V4L2_CID_DIGITAL_GAIN:
+		ret = cci_write(ov64b40->cci, OV64B40_REG_MEC_LONG_DIG_GAIN,
+				ctrl->val << 6, NULL);
+		break;
+	case V4L2_CID_VBLANK: {
+		int vts = ctrl->val + ov64b40->mode->height;
+
+		cci_write(ov64b40->cci, OV64B40_REG_TIMINGS_VTS_LOW, vts, &ret);
+		cci_write(ov64b40->cci, OV64B40_REG_TIMINGS_VTS_MID,
+			  (vts >> 8), &ret);
+		cci_write(ov64b40->cci, OV64B40_REG_TIMINGS_VTS_HIGH,
+			  (vts >> 16), &ret);
+		break;
+	}
+	case V4L2_CID_VFLIP:
+		ret = cci_update_bits(ov64b40->cci, OV64B40_REG_TIMING_CTRL_20,
+				      OV64B40_TIMING_CTRL_20_VFLIP,
+				      ctrl->val << 2,
+				      NULL);
+		break;
+	case V4L2_CID_HFLIP:
+		ret = cci_update_bits(ov64b40->cci, OV64B40_REG_TIMING_CTRL_21,
+				      OV64B40_TIMING_CTRL_21_HFLIP,
+				      ctrl->val ? 0
+						: OV64B40_TIMING_CTRL_21_HFLIP,
+				      NULL);
+		break;
+	case V4L2_CID_TEST_PATTERN:
+		ret = cci_write(ov64b40->cci, OV64B40_REG_TEST_PATTERN,
+				ov64b40_test_pattern_val[ctrl->val], NULL);
+		break;
+	default:
+		dev_err(ov64b40->dev, "Unhandled control: %#x\n", ctrl->id);
+		ret = -EINVAL;
+		break;
+	}
+
+	if (pm_status > 0) {
+		pm_runtime_mark_last_busy(ov64b40->dev);
+		pm_runtime_put_autosuspend(ov64b40->dev);
+	}
+
+	return ret;
+}
+
+static const struct v4l2_ctrl_ops ov64b40_ctrl_ops = {
+	.s_ctrl = ov64b40_set_ctrl,
+};
+
+static int ov64b40_init_controls(struct ov64b40 *ov64b40)
+{
+	int exp_max, hblank_val, vblank_max, vblank_def;
+	struct v4l2_ctrl_handler *hdlr = &ov64b40->ctrl_handler;
+	struct v4l2_fwnode_device_properties props;
+	const struct ov64b40_timings *timings;
+	int ret;
+
+	ret = v4l2_ctrl_handler_init(hdlr, 11);
+	if (ret)
+		return ret;
+
+	ov64b40->pixel_rate = v4l2_ctrl_new_std(hdlr, &ov64b40_ctrl_ops,
+						V4L2_CID_PIXEL_RATE,
+						ov64b40_modes[0].pixel_rate,
+						ov64b40_modes[0].pixel_rate, 1,
+						ov64b40_modes[0].pixel_rate);
+
+	v4l2_ctrl_new_std_menu_items(hdlr, &ov64b40_ctrl_ops,
+				     V4L2_CID_TEST_PATTERN,
+				     ARRAY_SIZE(ov64b40_test_pattern_menu) - 1,
+				     0, 0, ov64b40_test_pattern_menu);
+
+	timings = &ov64b40_modes[0].timings_default;
+	exp_max = timings->vts - OV64B40_EXPOSURE_MARGIN;
+	ov64b40->exposure = v4l2_ctrl_new_std(hdlr, &ov64b40_ctrl_ops,
+					      V4L2_CID_EXPOSURE,
+					      OV64B40_EXPOSURE_MIN, exp_max, 1,
+					      OV64B40_EXPOSURE_MIN);
+
+	hblank_val = timings->ppl * 8 - ov64b40->mode->width;
+	ov64b40->hblank = v4l2_ctrl_new_std(hdlr, &ov64b40_ctrl_ops,
+					    V4L2_CID_HBLANK, hblank_val,
+					    hblank_val, 1, hblank_val);
+	if (ov64b40->hblank)
+		ov64b40->hblank->flags |= V4L2_CTRL_FLAG_READ_ONLY;
+
+	vblank_def = timings->vts - ov64b40->mode->height;
+	vblank_max = OV64B40_VTS_MAX - ov64b40->mode->height;
+	ov64b40->vblank = v4l2_ctrl_new_std(hdlr, &ov64b40_ctrl_ops,
+					    V4L2_CID_VBLANK, OV64B40_VBLANK_MIN,
+					    vblank_max, 1, vblank_def);
+
+	v4l2_ctrl_new_std(hdlr, &ov64b40_ctrl_ops, V4L2_CID_ANALOGUE_GAIN,
+			  OV64B40_ANA_GAIN_MIN, OV64B40_ANA_GAIN_MAX, 1,
+			  OV64B40_ANA_GAIN_DEFAULT);
+
+	v4l2_ctrl_new_std(hdlr, &ov64b40_ctrl_ops, V4L2_CID_DIGITAL_GAIN,
+			  OV64B40_DIG_GAIN_MIN, OV64B40_DIG_GAIN_MAX, 1,
+			  OV64B40_DIG_GAIN_DEFAULT);
+
+	ov64b40->hflip = v4l2_ctrl_new_std(hdlr, &ov64b40_ctrl_ops,
+					   V4L2_CID_HFLIP, 0, 1, 1, 0);
+	if (ov64b40->hflip)
+		ov64b40->hflip->flags |= V4L2_CTRL_FLAG_MODIFY_LAYOUT;
+
+	ov64b40->vflip = v4l2_ctrl_new_std(hdlr, &ov64b40_ctrl_ops,
+					   V4L2_CID_VFLIP, 0, 1, 1, 0);
+	if (ov64b40->vflip)
+		ov64b40->vflip->flags |= V4L2_CTRL_FLAG_MODIFY_LAYOUT;
+
+	if (hdlr->error) {
+		ret = hdlr->error;
+		dev_err(ov64b40->dev, "control init failed: %d\n", ret);
+		goto error_free_hdlr;
+	}
+
+	ret = v4l2_fwnode_device_parse(ov64b40->dev, &props);
+	if (ret)
+		goto error_free_hdlr;
+
+	ret = v4l2_ctrl_new_fwnode_properties(hdlr, &ov64b40_ctrl_ops,
+					      &props);
+	if (ret)
+		goto error_free_hdlr;
+
+	ov64b40->sd.ctrl_handler = hdlr;
+
+	return 0;
+
+error_free_hdlr:
+	v4l2_ctrl_handler_free(hdlr);
+	return ret;
+}
+
+static int ov64b40_identify(struct ov64b40 *ov64b40)
+{
+	int ret;
+	u64 id;
+
+	ret = cci_read(ov64b40->cci, OV64B40_REG_CHIP_ID, &id, NULL);
+	if (ret) {
+		dev_err(ov64b40->dev, "Failed to read chip id: %d\n", ret);
+		return ret;
+	}
+
+	if (id != OV64B40_CHIP_ID) {
+		dev_err(ov64b40->dev, "chip id mismatch: %#llx\n", id);
+		return -ENODEV;
+	}
+
+	dev_dbg(ov64b40->dev, "OV64B40 chip identified: %#llx\n", id);
+
+	return 0;
+}
+
+static int ov64b40_parse_dt(struct ov64b40 *ov64b40)
+{
+	struct v4l2_fwnode_endpoint v4l2_fwnode = {
+		.bus_type = V4L2_MBUS_CSI2_DPHY
+	};
+	struct fwnode_handle *endpoint;
+	int ret;
+
+	endpoint = fwnode_graph_get_next_endpoint(dev_fwnode(ov64b40->dev),
+						  NULL);
+	if (!endpoint) {
+		dev_err(ov64b40->dev, "Failed to find endpoint\n");
+		return -EINVAL;
+	}
+
+	ret = v4l2_fwnode_endpoint_alloc_parse(endpoint, &v4l2_fwnode);
+	fwnode_handle_put(endpoint);
+	if (ret) {
+		dev_err(ov64b40->dev, "Failed to parse endpoint\n");
+		return ret;
+	}
+
+	if (v4l2_fwnode.bus.mipi_csi2.num_data_lanes != 4) {
+		dev_err(ov64b40->dev, "Unsupported number of data lanes: %u\n",
+			v4l2_fwnode.bus.mipi_csi2.num_data_lanes);
+		v4l2_fwnode_endpoint_free(&v4l2_fwnode);
+		return -EINVAL;
+	}
+
+	v4l2_fwnode_endpoint_free(&v4l2_fwnode);
+
+	return 0;
+}
+
+static int ov64b40_get_regulators(struct ov64b40 *ov64b40)
+{
+	struct i2c_client *client = v4l2_get_subdevdata(&ov64b40->sd);
+	unsigned int i;
+
+	for (i = 0; i < ARRAY_SIZE(ov64b40_supply_names); i++)
+		ov64b40->supplies[i].supply = ov64b40_supply_names[i];
+
+	return devm_regulator_bulk_get(&client->dev,
+				       ARRAY_SIZE(ov64b40_supply_names),
+				       ov64b40->supplies);
+}
+
+static int ov64b40_probe(struct i2c_client *client)
+{
+	struct ov64b40 *ov64b40;
+	u32 xclk_freq;
+	int ret;
+
+	ov64b40 = devm_kzalloc(&client->dev, sizeof(*ov64b40), GFP_KERNEL);
+	if (!ov64b40)
+		return -ENOMEM;
+
+	ov64b40->dev = &client->dev;
+	v4l2_i2c_subdev_init(&ov64b40->sd, client, &ov64b40_subdev_ops);
+
+	ov64b40->cci = devm_cci_regmap_init_i2c(client, 16);
+	if (IS_ERR(ov64b40->cci)) {
+		dev_err(&client->dev, "Failed to initialize CCI\n");
+		return PTR_ERR(ov64b40->cci);
+	}
+
+	ov64b40->xclk = devm_clk_get(&client->dev, NULL);
+	if (IS_ERR(ov64b40->xclk))
+		return dev_err_probe(&client->dev, PTR_ERR(ov64b40->xclk),
+				     "Failed to get clock\n");
+
+	xclk_freq = clk_get_rate(ov64b40->xclk);
+	if (xclk_freq != OV64B40_XCLK_FREQ) {
+		dev_err(&client->dev, "Unsupported xclk frequency %u\n",
+			xclk_freq);
+		return -EINVAL;
+	}
+
+	ret = ov64b40_get_regulators(ov64b40);
+	if (ret)
+		return ret;
+
+	ov64b40->reset_gpio = devm_gpiod_get_optional(&client->dev, "reset",
+						      GPIOD_OUT_LOW);
+	if (IS_ERR(ov64b40->reset_gpio))
+		return dev_err_probe(&client->dev, PTR_ERR(ov64b40->reset_gpio),
+				     "Failed to get reset gpio\n");
+
+	ret = ov64b40_parse_dt(ov64b40);
+	if (ret)
+		return ret;
+
+	ret = ov64b40_power_on(&client->dev);
+	if (ret)
+		return ret;
+
+	ret = ov64b40_identify(ov64b40);
+	if (ret)
+		goto error_poweroff;
+
+	ov64b40->mode = &ov64b40_modes[0];
+
+	pm_runtime_set_active(&client->dev);
+	pm_runtime_get_noresume(&client->dev);
+	pm_runtime_enable(&client->dev);
+	pm_runtime_set_autosuspend_delay(&client->dev, 1000);
+	pm_runtime_use_autosuspend(&client->dev);
+
+	ret = ov64b40_init_controls(ov64b40);
+	if (ret)
+		goto error_poweroff;
+
+	/* Initialize subdev */
+	ov64b40->sd.internal_ops = &ov64b40_internal_ops;
+	ov64b40->sd.flags = V4L2_SUBDEV_FL_HAS_DEVNODE;
+	ov64b40->sd.entity.function = MEDIA_ENT_F_CAM_SENSOR;
+
+	ov64b40->pad.flags = MEDIA_PAD_FL_SOURCE;
+	ret = media_entity_pads_init(&ov64b40->sd.entity, 1, &ov64b40->pad);
+	if (ret) {
+		dev_err(&client->dev, "failed to init entity pads: %d\n", ret);
+		goto error_handler_free;
+	}
+
+	ov64b40->sd.state_lock = ov64b40->ctrl_handler.lock;
+	ret = v4l2_subdev_init_finalize(&ov64b40->sd);
+	if (ret < 0) {
+		dev_err(&client->dev, "subdev init error: %d\n", ret);
+		goto error_media_entity;
+	}
+
+	ret = v4l2_async_register_subdev_sensor(&ov64b40->sd);
+	if (ret < 0) {
+		dev_err(&client->dev,
+			"failed to register sensor sub-device: %d\n", ret);
+		goto error_subdev_cleanup;
+	}
+
+	pm_runtime_mark_last_busy(&client->dev);
+	pm_runtime_put_autosuspend(&client->dev);
+
+	return 0;
+
+error_subdev_cleanup:
+	v4l2_subdev_cleanup(&ov64b40->sd);
+error_media_entity:
+	media_entity_cleanup(&ov64b40->sd.entity);
+error_handler_free:
+	v4l2_ctrl_handler_free(ov64b40->sd.ctrl_handler);
+error_poweroff:
+	ov64b40_power_off(&client->dev);
+	pm_runtime_set_suspended(&client->dev);
+
+	return ret;
+}
+
+static void ov64b40_remove(struct i2c_client *client)
+{
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+
+	v4l2_async_unregister_subdev(sd);
+	v4l2_subdev_cleanup(sd);
+	media_entity_cleanup(&sd->entity);
+	v4l2_ctrl_handler_free(sd->ctrl_handler);
+
+	pm_runtime_disable(&client->dev);
+	if (!pm_runtime_status_suspended(&client->dev))
+		ov64b40_power_off(&client->dev);
+	pm_runtime_set_suspended(&client->dev);
+}
+
+static const struct of_device_id ov64b40_of_ids[] = {
+	{ .compatible = "ovti,ov64b40" },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, ov64b40_of_ids);
+
+static const struct dev_pm_ops ov64b40_pm_ops = {
+	SET_RUNTIME_PM_OPS(ov64b40_power_off, ov64b40_power_on, NULL)
+};
+
+static struct i2c_driver ov64b40_i2c_driver = {
+	.driver	= {
+		.name = "ov64b40",
+		.of_match_table	= ov64b40_of_ids,
+		.pm = &ov64b40_pm_ops,
+	},
+	.probe	= ov64b40_probe,
+	.remove	= ov64b40_remove,
+};
+
+module_i2c_driver(ov64b40_i2c_driver);
+
+MODULE_DESCRIPTION("OmniVision OV64B40 sensor driver");
+MODULE_LICENSE("GPL");
